<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML>
<head>
<!-- official OU colors:  crimson #4F1315 cream #FDF908  source: https://alumni.ou.edu/home/webcomm/webguide/designconsiderations.html -->
<!-- off white #f5f4f0   reddish slate #E4D2C1  OU brickred #4F1315   light green #FFEFB2  light blue #F6F8FF   light-green #fceeaf   royal blue #0D2A70   light-green #FFFCCC  light gray F6F4F0 -->
<!--<body bgcolor=#FDF9D8  style="font-family:arial;font-size:15px"> -->
<link rel="stylesheet" href="style.css">
<style type="screen.css">
    <style>
      a {
        text-decoration: none;
      }
      a:link {
        color: #6b5b95;
      }
      a:visited {
        color: #feb236;
      }
      a:hover {
        color: #d64161;
      }
</style>

<script type="text/javascript" async
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<!-- Format code blocks. Must use code tag inside of pre tags to get the block effect-->
<style type="text/css">
	pre code {
	  background-color: #eee;
	  border: 1px solid #999;
	  display: block;
	  padding: 20px;
	}
</style>


<!-- Format the vertical spacing in lists. 	    border:1px solid grey;-->
<style type="text/css">
	li{
	    margin-top: 10px;
	}

	li:first-child {
	    margin-top:0;
	}
</style>

<!-- Format table -->

<style>
<!--
table.table1 {
  border:1px solid #A95AA1;
  width:100%;
  table-layout:fixed;
}

table.table1 caption {
	font-family: Arial, Helvetica, sans-serif;
    text-align: center;
    margin-bottom: 5px;
    font-size: 240%;
    padding: 5px;
    font-weight: bold;
}


table.table1 tbody>tr>:nth-child(1){
 color:#F5793A;
 text-align:center;
 font-size: 2em;
 width: 40;
}

table.table1 tbody>tr>:nth-child(2){
 text-align:center;
 width: 200;
}

table.table1 tbody>tr>:nth-child(3){
 color:#0F2080;
 text-align:left;
 padding: 10px;
 font-size: 1.2em;
 word-wrap:break-word;
 width: 200;
}

table.table1 th { 
	font-size: 1.4em;
	border:1px solid  #A95AA1; 
	text-align:center;
}

table.table1 td{
	border:1px solid;
	color: #A95AA1; 
}



table.table2 {
    border:1px solid black;
    width:100%;
    table-layout:fixed;
    font-family: arial, sans-serif;
    border-collapse: collapse;
    width: 100%;
}

table.table2 tbody>tr>:nth-child(1){
 color:black;
 text-align:left;
 font-size: 1em;
 border: 1px solid black;
 text-align: left;
 padding: 8px;
 width: 50px;
}

table.table2 tbody>tr>:nth-child(2){
    color:black;
    text-align:left;
    font-size: 1em;
    border: 1px solid black;
    text-align: left;
    padding: 8px;
    width: 300px;
}

table.table2 th { 
	font-size: 1.4em;
	border:1px solid black; 
	text-align:center;
}

table.table2 tr:nth-child(even) {
  background-color: #dddddd;
}

-->
</style>


<!--
table.table2 td, th {
  border: 1px solid  #A95AA1;
  text-align: left;
  padding: 8px;
}
-->



<style type="text/css">
<!--
 .tab { margin-left: 25px; }
-->
</style>


<!-- tab a section -->
<STYLE TYPE="text/css">
<!--
.indented
 {
 padding-left: 50pt;
 padding-right: 50pt;
 }
-->
</STYLE>

<!-- tab a section -->
<STYLE TYPE="text/css">

ul {
  list-style-position: inside;
  padding-left: 0;
}
</STYLE>

<!-- A nice ligth grey: bgcolor=#f5f4f0  -->
</head>
<body style="font-family:Arial, Helvetica, sans-serif; font-size:15px">


<h1> <em>pymolsnips:</em> PyMOL Script Writing with Code Templates</h1>

<h2> <a name="FASTLINKS">Quick links for the impatient</a></h2>

<ul>
  <li> <a href="#gallery"> Gallery of examples</a></li>
  <li> <a href="#FASTLINKS3"> Snippet categories</a></li>
  <li> <a href="#editors"> List of supported text editors</a></li>
  <li> <a href="#install"> Demos and installation instructions by text editor</li>
  <li> <a href="#Snippet Lists">List of snippets by category and tab trigger name</a></li>
</ul>


<p>
<a href="https://pymol.org/2/">PyMOL</a> is a leading molecular graphics program for making images of proteins and nucleic acids for publication.
PyMOL's vast array of parameters provides exquisite control over the appearance of the output.
PyMOL is often used to make cover images for <a href="https://pymolwiki.org/index.php/Main_Page">scientific journals</a>.
PyMOL is also popular for making movies of molecules.
</p>

<p>
The PyMOL GUI is useful for making the images of global scenes, but PyMOL rapidly becomes tedious to use to make images of detailed scenes.
The PyMOL macro language (pml) can be used to set parameter values and execute commands to make customized scenes of biomolecules in PyMOL's viewport.
These scenes can be made into static images for posters, seminars, and manuscripts, or they can serve as parts of molecular movies.
The macro language sends arguments to Python functions, but its syntax is simpler for non-programmers to understand than the syntax of Python code.
</p>

<p>
Over 100 lines of pml commands and settings are required to make more sophisticated images.
It is difficult to issue so many commands through PyMOL's graphical user interface (GUI) without making mistakes.
If the commands are not saved to an open script file with a <em>.pml</em> file extension, to an opened log file, or to a frequently saved session file, the work can be lost. 
(Use the <b>spse</b> function in the <em>pymolshortcuts.py</em> file in the <a href="https://github.com/MooersLab/pymolshortcuts">pymolshortcuts</a> repository to save session files with time stamps to avoid overwriting previously saved session files.)
</p>

<p>
Here are of some figures that are impossible or tedious to make via the PyMOL GUI alone.
</p>
<!-- ![Gallery](https://github.com/MooersLab/pymolsnips/blob/master/images/Gallery.png?raw=true "Gallery") -->

<link rel="stylesheet" href="style.css">


<a name="gallery"> </a>
    <table class="table1">
		<caption> Gallery </caption>
       <tr>
        <th WIDTH="40">Tab trigger</th>
        <th WIDTH="200">Output from code </th>
        <th WIDTH="100"><b>Category</b> and Description </th>
      </tr>  
      <tr>
        <td> ao</td>
        <td><img src="./images/ao.png" alt="HTML5 Icon" style="width:600px;height:600px;"> </td>
        <td><b>ambient occlusion</b> <p>Applies a photorealistic effect. The effect is not available via a GUI pulldown mean. This code can be applied to any molecule in the sphere representaton. It applies ray tracing so moving the molecule in the viewport ruins the effect. Enter <b>hide spheres</b> to remove the spheres.</p> </td>
     </tr>
      <tr>
	<!-- pdb code 5d99, 27-mer RNA hairpin -->
        <td>aod</td>
        <td align="center"><img src="./images/5d99AOD.png" alt="HTML5 Icon" style="width:468px;height:741px;"> </td>
        <td align="left"><h4>ambient occlusion</h4> <p>This variant of ambient occulusion colors the carbon atoms black. This code can be applied to any molecule in the sphere representation.</p> </td>
     </tr>
     <tr>
	<!-- pdb code 5d99, 27-mer RNA hairpin -->
        <td>aobw</td>
        <td><img src="./images/AOBWdsRNA.png" alt="HTML5 Icon" style="width:364px;height:630px;"> </td>
        <td><h4>ambient occlusion</h4> <p>This is the grayscale verison of ambient occlusions. Grayscale is not an option available through a pulldown in PyMOL. This code can be applied to any molecule in the sphere representation.</p> </td>
     </tr>
     <tr>
	<!-- pdb code 3nd3 -->
        <td>aodbw</td>
        <td align="center"><img src="./images/U8AODBW.png" alt="HTML5 Icon" style="width:459px;height:874px;"> </td>
        <td align="left"><h4>ambient occlusion</h4> <p>This variant of ambient occulusion colors the carbon atoms black and then applies grayscale. Grayscale is not an option available through a pulldown in PyMOL. This code can be applied to any molecule in a sphere representation. (PBD: code 3nd3).</p> </td>
     </tr>
      <tr>
        <td>bw</td>
        <td align="center"><img src="./images/bw.png" alt="HTML5 Icon" style="width:300px;height:300px;"> </td>
        <td align="left"><h4>black and white cartoon</h4> <p>This code can be applied to any molecule in a sphere, surface, or cartoon representation.</p></td>
     </tr>
      <tr>
        <td>carved</td>
        <td align="center"><img src="./images/carved.png" alt="HTML5 Icon" style="width:300px;height:246px;"> </td>
        <td align="left"><h4>carved electron density</h4> <p>A 2Fo-Fc electron density map is carved around a glycan.</p></td>
     </tr>
      <tr>
        <td>coordinate</td>
        <td align="center"><img src="./images/coordinate.png" alt="HTML5 Icon" style="width:300px;height:224px;"> </td>
        <td align="left"><h4>coordinate covalent bonds</h4> <p>Example of specifying coordinate covalent bonds between RNA and a sodium (colored purple). The coordinate covalent bonds are not diplayed by default in PyMOL; each atom pair in a coordinate covlent bond has to be specficied. The H-bonds were also specifid. This snippet can be used as a template for the commands needed to display coordinate covalent bonds and H-bonds.</p> </td>
     </tr>
      <tr>
        <td>filledring</td>
        <td align="center"><img src="./images/filledring.png" alt="HTML5 Icon" style="width:284px;height:300px;"> </td>
        <td align="left"><h4>filled-ring cartoon for nucleic acids</h4> <p>Filled ring cartoon for nucleic acids with the backbone highlighted by a flat ribbon.</p> </td>
     </tr>
      <tr>
        <td>sc222</td>
        <td align="center"><img src="./images/sc222.png" alt="HTML5 Icon" style="width:300px;height:200px;"> </td>
        <td align="left"><h4>Generate 2 x 2 x 2 array of unit cells with symmetry mates.</h4> <p>Run Thomas Holder's script supercell.py to generate 2 x 2 x 2 array of unit cells with symmetry mates filling the unit cells.</p> </td>
     </tr>
      <tr>
        <td>stack</td>
        <td align="center"><img src="./images/stack.png" alt="HTML5 Icon" style="width:300px;height:166px;"> </td>
        <td align="left"><h4>Base stacking</h4> <p>Base stacking figure for a pair of base pairs with the top base pair colored dark to enhance the depth perception. The major groove is on the top.</p> </td>
     </tr>  
</table>

<p>
It is challenging to recall the pml syntax when you are not using PyMOL every day, which is the situtation for most PyMOL users.
One solution to this problem is to use a library of code fragments, called <b>snippets</b> or <b>templates</b>, to build a script in a text editor.
</p>

<h3>Brief Notes on Installation and Updates</h3>

<p>
Users may want to update their copy of the library as new snippets are added to the library on GitHub.
The date of the last update and its nature are recorded in the GitHub repository.
GitHub has only experimental support for downloading part of a repository.
In the meantime, users must download the entire repository either as a zip file via the <b>code &rarr; download zip</b> pull-down menu or by entering the command <b>git clone https://github.com/MooersLab/pymolsnips.git</b> if they have git on their computer.
The second option enables subsequent updating of the library via the <b>git pull origin master</b> command.
</p>

<h3>Navigating this document</h3>

<p>
This document has over 12,000 words.
You do not need to read much of it to get started.
Just go to the section below about your favorite text editor to learn how to learn the snippet library.
</p>

<p>
This long document has a large number of hyperlinks that enable rapid movement to desired sites below and the return to the menus at the top of the file.
The hyperlinks reduce the need to scroll through the document.
Some of the text is hidden from view and has to be toggled or unfolded to expose it.
</p>

<p>
If you are considering switching text editors or have never used one, then you should invest an hour in reading the synopses of the 18 supported text editors.
This can serve as a guide to selecting a text editor.
</p>

<p>	
Yes, we know.
We have too much information here for a standard README.md file.
We plan to move much of the detailed information below to a GitHub Pages webpage shortly.
</p>



<p>
<em>pymolsnips</em> is a library of pml code fragments that have been formatted for several popular <em>text editors</em>.
Note the word processors (e.g., MS Word, Libre Office, Open Office) are not included here: <em>word processors should never be used to store computer code!</em>
If you do so, you can inadvertently pick up hidden characters that are hard to find and remove during debugging.
In addition, word processing documents are usually binary files that are difficult to put under version control.
</p>

<p>
Text editors have tools for supporting the editing of computer code files.
Although these files have many different file extensions that map them to the programming language of the code that they contain, they are all plain text files.
One of the editing tools available in text editors is support for the retrieval and insertion of computer code fragments (i.e, snippets) to save time and reduce errors while assembling a new script file.
</p>

<p>
Protocols for snippet installation for each text editor are found <a href="#install"> below</a>.
Some editors take snippet libraries in one file while other editors require that each snippet resides in a separate file.
Most editors have their own format for snippets.
The user downloads the file or folder of snippet files for their text editor and then installs the snippets according to the requirements for a particular text editor.
The one exception is the CudaText editor.
The snippets come pre-installed with this editor.
</p>

<p>
Each code fragment has a unique name that also serves as a tab trigger.
The user enters the name of the code snippet and hits the <b>tab</b> key to insert the lines of code.
Tab stops exist at sites where parameters can be edited.
Tab stops are mirrored when the parameters are identical.
Mirroring enable the simultaneous editing of these sites to reduce the chance of overlooking a site that needed editing.
</p>

<p>
The animation below demonstrates the use of the <b>ao</b> tab trigger in <em>Visual Studio Code</em> to insert 16 lines of code for generating the ambient occlusion effect.
You can learn how this is done <a href="#VisualStudioCode"> here </a>.
Two-levels of cascading menus appear.
The menu on the left shows the alternate tab triggers that contain the letters <b>a</b> and <b>o</b>.
The corresponding code for the selected tab trigger is displayed in the right window.
By entering these two letters, you have inserted 16 lines of code!
</p>

<p align="center"><img src="gifs/VSCaoSnip.gif"></p>

<p>
One thing to note when using tab stops is that the mirrored selections are very <em>fragile</em>:
Before you begin typing, make sure that the mirrored sections are all highlighted.
It is easy to exit the tab stop selections by the click of a mouse or even a keyboard movement!
Getting your selections re-highlighted is not difficult, but the required steps vary between text editors.
Find more information about mirrored tab stops in the installation instructions for each text editor.
</p>

<p>
The animation below demonstrates the use of mirrored tab stops where changed default values are mirrored at identical sites.
If mirrored tab stops are available for a text editor, you can learn more about them in the installation instructions of that text editor.
</p>

<p align="center"><img src="gifs/VSCMirror.gif"></p>

<a href=#FASTLINKS>Return to quick links section at top</a>
<hr>

<h2>  <a name="categories"> Snippet categories: </a> </h2>
<p>
Each snippet is described below.
The snippets are grouped into categories.
Hyperlinks will take you to a particular category.
</p>

<a name="FASTLINKS3"></a>


<ul>
<li><a href="#Alternate locators"> Alternate locators </a></li>
<li><a href="#Change orientation"> Change orientation </a></li>
<li><a href="#Color scheme"> Color scheme </a></li>
<li><a href="#Electron density"> Electron density </a></li>
<li><a href="#FileIO"> FileIO</a></li>
<li><a href="#H-bonds">H-bonds</a></li>
<li><a href="#Help">Help</a></li>
<li><a href="#Jupyter">Jupyter</a></li>
<li><a href="#Labels">Labels</a></li>
<li><a href="#List Snippets">List Snippets</a></li>
<li><a href="#Measurements">Measurements</a></li>
<li><a href="#Molecular representation">Molecular representation</a></li>
<li><a href="#Nucleic Acids">Nucleic Acids</a></li>
<li><a href="#Objects">Objects</a></li>
<li><a href="#Print">Print</a></li>
<li><a href="#Programming">Programming</a></li>
<li><a href="#Pymolrc">Pymolrc</a></li>
<li><a href="#Selection">Selection</a></li>
<li><a href="#Settings">Settings</a></li>
<li><a href="#Stereo">Stereo</a></li>
<li><a href="#Trajectories">Trajectories</a></li>
<li><a href="#Workshop">Workshop</a></li>
</ul>

<a href=#FASTLINKS>Return to quick links section at top</a>

<hr>


<h2>  <a name="editors">Supported text editors:</a> </h2>
<a name="FASTLINKS2"></a>
<p>
PyMOL scripts can be coded in electronic notebook, integrated development environments, and text editors.
All three types of coding platforms support the use of snippets to varying degrees.
Here, we foucs on text editors because the support for editing *pml* code is the strongest with this category of coding environment.
</p>


<p>
Note that some text editors use the same library of snippets.
For example, <em>Emacs</em> and <em>spacemacs</em> can use the same snippet library that is managed by the yasnippets package.
Likewise, <em>Vim</em> and <em>neovim</em> share some of the same plugin managers and snippet libraries.
There are at least three kinds of snippet systems available for <em>Vim</em> and <em>neovim</em>.
</p>


<ul>
<li> <a href="#ATOM"> Atom </a></li>
<li><a href="#bbedit"> BBEdit (limited to Mac OS)</a></li>
<li><a href="#brackets"> Brackets </a></li>
<li> <a href="#cudatext"> CudaText </a></li>
<li> <a href="#emacs"> Emacs (uses <a href="#yasnippets">yasnippets</a> package)</li>
<li> <a href="#espresso"> Espresso </a></li>
<li> <a href="#geany"> Geany </a></li>
<li> <a href="#gedit"> Gedit </a></li>
<li> <a href="#kate"> Kate </a></li>
<li> <a href="#komodo"> Komodo Edit </a></li>
<li> <a href="#LightTable"> Light Table </a></li>
<li> <a href="#micro"> Micro </a></li>
<li> <a href="#Neovim"> Neovim</a> uses the following to manage snippet plugins</li> 
  <ol>
    <li><a href="#Ultisnips">Ultisnips</a></li>
    <li><a href="#Neosnippets">Neosnippets</a></li>
    <li><a href="#Snipmate">Snipmate</a> plugins to manage snippets) </a></li>
  </ol>
<li> <a href="#spacemacs"> Spacemacs (uses yasnippets)</a></li>
<li> <a href="#SublimeText3"> Sublime Text 3 </a></li>
<li> <a href="#TextMate"> TextMate (limited to Mac OS)</a></li>
<li> <a href="#Vim"> Vim </a>(uses Ultisnips, Neosnippets, or Snipmate plugins to manage snippets)</li>
  <ol>
    <li><a href="#Ultisnips">Ultisnips</a></li>
    <li><a href="#Neosnippets">Neosnippets</a></li>
    <li><a href="#Snipmate">Snipmate</a> plugins to manage snippets) </a></li>
  </ol>
<li> <a href="#VisualStudioCode"> Visual Studio Code </a></li>
</ul>


<a href=#FASTLINKS2>Return to list of editors above.</a ><br>
<a href=#FASTLINKS>Return to the quick links section at top.</a>

<p>
If you are considering switching editors, we recommend Visual Studio Code (VSC).
You do have to install plugins, but the process of doing so is quite painless.
Our second tier of text editors includes would be Sublime Text, Textmate, and Atom.
</p>

<p>
You might also consider <em>Geany</em>.
It is very lightweight, very fast, and very easily configurable.
It is a good editor if you care about agility.  
See the following for second opinions on the best Python editor <a href="https://www.slant.co/topics/18408/~python-ides-or-editors-for-beginners">for beginners</a> and <a href="https://www.slant.co/topics/366/~best-python-ides-or-editors"> for all users </a>. 
Note that the <em>thonny</em> editor that was recommended for beginners is more of a training tool than a productive tool.
It does not support code snippets, yet.
</p>

<p>
If your favorite editor is not listed, please post an issue <a href="https://github.com/MooersLab/pymolsnips/issues">here</a>.
I will be notified immediately by e-mail and will try to develop a snippet library for the requested editor.
</p>

<p>
Note that some editors that are available as binaries only for Windows like <em>Notepad++</em> can be run on Mac OS or Linux by using wine or wine bottler.
</p>

<p>
Some of these text editors can take hours to customize to fit your needs; however, you only need to know about 5% of the options to become productive with these editors.
</p>

<h1> <a name="install"></a>Installaton instructions by editor</h1>
 
 

<a  name='ATOM'><h2> Atom (Universal) </h2></a>

<p>
<a href="https://atom.io"><b>Atom</b></a> is a favorite of professional programmers because it is easily extended and customized.
Atom advertises itself to be the "hackable text editor".
It integrates with GitHub nicely, and the GUI is attractive.
</p>

<p>
<em>Atom's</em> startup speed bogs down as more plugins are added.
The work-around is to keep <em>Atom</em> open all of the time.
</p>

<p>
There is an extensive collection of plugins available for <em>Atom</em>.
The installation and updating of plugins can be slow compared to other editors.
</p>

<h3>Installing and using snippets with Atom</h3>

<p>
You will need the <em>snippets</em> package to be able to use the pymolpy snippets.  
The package installer is very intuitive.
Go to <b>Packages &rarr; Settings View &rarr; Install packages/themes</b>.
Search for <em>snippets</em> and click the install button.
It may already be installed, but you must make sure that the snippets package is enabled (green bar along the button).
</p>

<p>
Shown below is an enabled <em>snippets</em> package.
</p>

<p align="center"><img src="gifs/atomSnippetsEnable.gif"></p>

<p>
Others have developed a PyMOL lexer for *Atom* so that you can enjoy syntax highlighting.
Go to <b>Packages &rarr; Settings View &rarr; Install packages/theme</b> and search for <em>.language-pymol</em>.
Click install.
</p>

<p align="center"><img src="gifs/atomPyLanguageInstall.gif"></p>

<p>
The snippets for all programming languages are stored in a single file that is called <em>snippets.cson</em>.
The snippets for different languages are separated by the first line of a snippet library for a specific language.
This line contains a specification of the scope.
That is the kinds of script files to which a set of snippets applies.
The <em>snippets.cson</em> file is stored in a hidden folder on your home directory called <em>~.atom/snippets.cson</em>.
</p>

<p>
You can concatenate <a href="https://github.com/MooersLab/pymolsnips/tree/master/atompymolsnips">this</a> file of PyMOL snippets for <em>Atom</em> to your existing <em>snippets.cson</em> file.
</p>

<p>	
To do this, Mac users enter: <code>cp -a ~/pymolsnips/atompymolsnips/pymolsnippets.cson ~/.atom/snippets.cson</code>
</p>
	
<p>	
You can also access the <em>snippets.cson</em> file via the <b>Welcome Guide</b> of <b>Atom</b>, under the section called <b>Add a Snippet &rarr; Open your snippets</b>.
Or, you can access the <em>snippets.cson</em> file through <b>Atom (menu) &rarr; Snippets...</b>.
</p>
	
<p>	
For windows users, it might be easiest to access the <em>snippets.cson</em> file via the <em>Welcome Guide</em>. 
Paste the <em>snippets.cson</em> provided <a href="https://github.com/MooersLab/pymolsnips/tree/master/atompymolsnips">here</a> into this file. 
</p>
	
<p>	
A third option available for windows users, is to navigate to the <em>.atom</em> folder through the GUI of <b>File Explorer</b> and open the <em>snippets.cson</em> file.
</p>
	
<p>	
Below is an example of the <em>threeMaps</em> snippet and a demonstration of its mirrored tab stops.
</p>
	
<p align="center"><img src="gifs/AtomSnipsUse.gif"></p>

<a href=#FASTLINKS2>Return to list of editors above.</a >


<a name='bbedit'><h2> BBEdit (Mac only) </h2></a>

<p>
<a href="https://www.barebones.com/products/bbedit/index.html"><b>BBEdit</b></a> requires a license and a one-time fee.
Major upgrades require additional fees. You will need macOS 10.14.2 or later. 
</p>
	
<p>	
The snippet system for <em>BBedit</em> is simple and elegant.
The snippets are stored as clippings with one snippet or clipping per file.
The PyMOL pml snippets end with the file extension <em>*.pml</em>.
The clippings can be stored in folders by language.
</p>
	
	
<h3>Installing and using snippets with BBEdit</h3>

<p>
Users can create nested subfolders for groups of related snippets.
The folders of clippings are stored in <code>~/Library/Application\ Support/BBEdit/Clippings</code>. 
You will need to add <a href="https://github.com/MooersLab/pymolsnips/tree/master/bbeditpymolsnips">this</a> snippets folder to your BBEdit clippings.
</p>
	
<p>	
To do this, enter <code>cp -a ~/pymolsnips/bbeditpymolsnips ~/Library/Application\ Support/BBEdit/Clippings/</code>.
</p>
	
<p>	
The disadvantages of these clippings include the lack of tab triggers and tab stops.
There is also presently the lack of a PyMOL lexer for syntax highlighting.
</p>
	
	
<a href=#FASTLINKS2>Return to list of editors above.</a >



<a name='brackets'><h2> Brackets (Universal) </h2></a>

<p>
The development of <a href="http://brackets.io"><em>Brackets</em></a> is lead by a team at Adobe.
*Brackets* was designed for web developers with a focus on HTML, CSS, and JavaScript.
However, <em>Brackets</em> now has support for a large number of languages.
<em>Brackets</em> provides a Quick Edit and a Live Preview mode that run simultaneously, so changes in the <em>pml</em> code are dispayed immediately in the preview window.
</p>

<p>	
<em>Brackets</em> has icons, located on the right panel, for navigating to the interactive settings on the application.
The zig-zag line launches the Live Preview.
The lego piece launches the extension manager.
Once you install an extension, a light bulb icon will appear that launches the <b>Snippets Manager</b>.
</p>
	
	
<h3>Installing and using snippets with Brackets</h3>

<p>
Go to <b>File &rarr; Extension Manager</b> or press the lego piece in the right panel.
Search for <b>Brackets Snippets (by edc)</b> and install it.
This is the snippet manager.
It has a GUI that enables the manual creation of user-defined snippets.
</p>
	
	
<p align="center"><img src="gifs/bracketsExtInstall.gif"></p>
<p>
The GUI should automatically refresh, but if not, close and reopen *Brackets*.
The light bulb icon will now appear in the right panel.
Click on the light bulb to open the <b>Snippets Manager</b>.
</p>
	
<p>	
The <b>settings</b> tab opens a menu with an import button.
Click <b>import</b> and choose the <a href="https://github.com/MooersLab/pymolsnips/tree/master/bracketspymolsnips"><em>bracketspymolsnips.yml</em></a> file that is available at the top of this page.
You have to download the file to your computer.
</p>
	
<p>	
All of the snippets for PyMOL are in this single file.
The next step is to choose the import scheme.
I chose the recommended option to skip snippets with the same tab trigger name.
</p>
	
<p align="center"><img src="gifs/bracketsSnipsInstall.gif"></p>

<p>
Until the PyMOL language is available via a lexer for <em>Brackets</em>, use Python for the scope of the pml files.
The Python lexer will provide some syntax highlighting.
This means that the PyMOL file needs to have a file extension of <em>.py</em> while it is being edited in *Brackets*.
When finished, save this file with the <em>.pml</em> extension to be able to use it in PyMOL.
</p>
	
<p>	
In the example of snippet use shown below, notice how all snippets that begin with <b>a</b> are suggested.
Only when I press <b>o</b> do the suggestions narrow.
This is helpful when you cannot remember the short-hand for a snippet.
</p>
	
<p align="center"><img src="gifs/bracketsSnipsUse.gif"></p>

<p>
Unfortunately, <em>Brackets</em> does not currently support tab stops.
</p>
	
<a href=#FASTLINKS2>Return to list of editors above.</a>


<a name='cudatext'><h2> CudaText (Universal)</h2></a>

<p>
<a href="http://uvviewsoft.com/cudatext/"><em>CudaText</em></a> is a free, cross-platform editor that is written in Lazarus, a variant of Pascal.
<em>Cudatext</em> is open source.
Its predecessor was <em>SynWrite</em>, which is no longer supported.
<em>CudaText</em> has numerous plugins available to it.
A PyMOL lexer and PyMOL snippets are available through the <em>CudaText</em> add-ons manager (thank you Alexey T.!).
The documentation for <em>CudaText</em> is located <a href="http://wiki.freepascal.org/CudaText">here</a>.
</p>

<h3>Installing and using snippets with CudaText</h3>

<p>
<em>CudaText</em> uses Python3 based plugins. 
<em>CudaText</em> expects to find the Python3.X from <a href="https://www.python.org/downloads/release/python-374/">Python.org</a>.
I downloaded the 64-bit version of Python3.7.4 from Python.org and used the installer to install the binary in about five minutes.
This Python interpreter is installed in the <em>/Applications</em> directory on the Mac. 
When I started <em>CudaText</em>, I did not have to edit any configuration files.
The snippets are stored one per file.
The files have the extensions of <em>.cuda-snippet</em>. 
<em>CudaText</em> snippets have names (i.e., descriptions), ids (i.e., tab triggers), and markers (i.e., tab stops).
<em>CudaText</em> snippets are stored in the user's Library on the Mac: <b>./Library/Application\ Support/CudaText/data/snippets/Std.PML</b> folder.
</p>
	
<p>	
You may have to create the <em>Std.PML</em> folder by giving the following bash command: <code>$ mkdir ./Library/Application\ Support/CudaText/data/snippets/Std.PML/</code>
</p>
	
<p>	
When executing <em>CudaText</em> for the first time on Windows, it will prompt you with the need for the extraction of files.
Proceed with the extraction and choose where you want the files to be extracted.
I chose to store it under <code>C:\Users\YOURUSERNAME\AppData\Roaming\CudaText</code> where I created the folder CudaText.
</p>
	
<p>	
Create the <em>Std.PML</em> sub directory under <code>C:\Users\YOURUSERNAME\AppData\Roaming\CudaText\data\snippets</code>.
</p>
	
<p>	
Any snippet add-on will automatically be installed in this directory.
Paste the files (not the folder) of <em>cudatextpymolsnips</em> into the <em>Std.PML</em> folder.
</p>
	
<p>	
After starting <em>CudaText</em>, install the snippets plugin by navigating to the <b>Plugins &rarr; Addons Manager &rarr; Install</b> and search for <b>plugin: Snippets</b> and select it.
A pop-up window will appear asking you if you would like to install, click OK.
I could not screen capture this pop-up window.
</p>
	
<p align="center"><img src="gifs/CudaTextpluginInstall.gif"></p>

<p>
You will also need to use the same pull-down to install the <em>lexer: PyMOL</em> and the <em>snippets: PyMOL</em>.
Restart <em>CudaText</em> to see the changes. 
When editing snippets or plugins in <b>CudaText</b> you can also update them by going to <b>Plugins &rarr; Addons Manager &rarr; Update</b>.
<p>

<p>
<em>CudaText</em> allows you to edit mirrored sites for uniform editing.
The animation below shows how to edit one mirror location and travel to the other mirrored sites by hitting the Tab key.
<p>
	
	
<p align="center"><img src="gifs/cudatextMirror.gif"></p>

<a href=#FASTLINKS2>Return to list of editors above.</a>


<a name='emacs'><h2> Emacs (universal)</h2></a>

<p>
<a href="http://uvviewsoft.com/cudatext/"><em>Emacs</em></a> is a free, open-source, cross-platform editor that is written in <em>Emacs</em> lisp (elisp), a variant of LISP.
LISP was developed in the early 1960s to support work on artificial intelligence.
<em>Emacs</em> was initiated in 1976 by Richard M. Stallman and others.
<em>Emacs</em> has a leadership succession plan that has been put into action.
Richard Stallman stepped away from the project in 2008, and others took over the maintenance of the core program.
This plan will greatly extend the lifetime of this software project.
</p>
	
<p>	
<em>Emacs</em> has been around for 45 years.
According to Lindy's Law, the future life expectancy of a software program is proportional to its current age, so every additional period of survival implies a longer remaining life expectancy.
This law suggests that <em>Emacs</em> should be around for another 45 years.
</p>

<p>
Over 100 other variants of <em>Emacs</em> are available.
<em>Spacemacs</em> is a version that has a gentler learning curve in part because it can be driven by a menu of commands that is readily accessible.
<em>Spacemacs</em> allows the use of Vim key bindings, <em>Emacs</em> key bindings, or both.
<em>Spacemacs</em> does have its own keybindings that have to be mastered.
This leads to the problem of needing to translate <em>Spacemacs</em> to Gnu <em>Emacs</em> and back again when tapping into external <em>Emacs</em> resources.
</p>
	
<p>	
<em>Doom Emacs</em> and <em>Prelude Emacs</em> are additional off-shoots of <em>Emacs</em> that are suppose to be easier to adopt.
They are similar to <em>Spacemacs</em>.
</p>
	
<p>	
<em>SciMax</em> is a variant of <em>Emacs</em> that is being optimized to support scientific publication and literate programming.
</p>
	
<p>	
For several reasons, we recommend becoming productive in GNU <em>Emacs</em> first before trying the <em>Emacs</em> variants.
First, the on-line resources that support mastery of GNU <em>Emacs</em> are far greater than the resources that support the variants.
Second, the documentation for the variants assumes a significant amount of prior knowledge of GNU <em>Emacs</em>.
Third, these variants run on top of GNU <em>Emacs</em>.
</p>
	
<p>	
It takes one to two weeks of nibbling away at <em>Emacs</em> tutorials to become productive at text editing in <em>Emacs</em>.
There is a built-in tutorial in <em>Emacs</em> as well as many on-line tutorials.
A visually appealing tutorial is available <a href="https://www.gnu.org/software/emacs/tour/">here</a>.
</p>
	
<p>	
It takes several months to become competent in <em>Emacs</em>.
It takes a life-time to master <em>Emacs</em> hence <em>Emacs</em> is known as a <b>life-time editor</b>.
</p>
	
<p>	
The one danger of <em>Emacs</em> is that it is easy to spend countless of hours configuring the <em>init.el</em> file and adding additional packages.
Many beginners spend too much time adding too many packages, most of which they never use.
As a beginner, you should ease off editing your <em>init.el</em> file if it has grown to 500 lines in the first several weeks of using <em>Emacs</em>.
A good practice is to add one package at a time and master each added package before adding another new package.
</p>
	
<p>	
I find that editing of <em>init.el</em> file occurs in widely spaced bouts.
Once everything is working to your satisfaction, there may be long periods of productivity during which there is no compelling need to fiddle with the <em>init.el</em> file.
</p>
	
<p>	
Many <em>Emacs</em> users have shared their <em>init.el</em> files on GitHub.
These can provide inspiration for the beginner.
Some have complex configurations that are often too complex to be copied and applied by the beginner.
</p>
	
<p>	
Many of the add-on packages add functionalities that are not available in other text editors or that become available years later.
The vast group of active contributors to <em>Emacs</em> will probably keep it in the lead for a long time into the future.
</p>

<p>	
Like PyMOL, GNU <em>Emacs</em> was designed to be highly extensible.
This is a large part of its popularity.
Over 4600 packages have been developed by users over the years.
Several package managers, including the <a href="https://melpa.org">melpa</a>, greatly ease package installation.
However, manual installation is often quite simple.
</p>
	
<p>	
The interface to <em>Emacs</em> is either a X-terminal window or a simple GUI that can be made complex by opening many <em>buffers</em>, which are like windows.
The management and navigation of buffers is a skill that the beginner needs to master early.
</p>

<p>
<em>Emacs</em> is designed to enable mouse free work although some buffers require that selections be made by using the mouse.
<em>Emacs</em> is infamous for having 1800 key bindings involving multiple keys.
The key bindings make heavy use of the alt or meta key.
However, there is in-line documentation, and some packages provide autosuggestions that greatly ease the discovery of new key bindings.
You only really need to memorize several dozen key bindings to become productive in <em>Emacs</em>.
The remaining knowledge is easy to acquire on an <em>as-needed</em> basis.
</p>
	
<p>	
Vim users can quickly become productive in <em>Emacs</em> because the Vim key bindings are available through the add-on package called evil-mode.
Mastery of the Vim key bindings in Vim is initially painful because it is like learning how to type for the first time.
Here again, mastery of a few dozen key bindings will enable you to be productive in Vim and in Evil-mode in <em>Emacs</em>.
<em>Vim</em> key bindings are thought to bear more efficient for editing text than <em>Emacs</em> key-bindings, so their availability in <em>Emacs</em> greatly enhances a <em>Vim</em> user's productivity in </em>Emacs</em>.
</p>
	
<p>	
Mastery of <em>Vim</em> key-bindings is a wise investment because they are available in many text editors and IDEs.
They are even available in <em>Jupyter Notebooks</em>!
In addition, Vim is more widely distributed than <em>Emacs</em>.
</p>

<p>
If you are a user of national laboratory computer systems, be cautious with your fantasy of being able to transfer your <em>init.el</em> file and reestablish your <em>Emacs</em> computing environment.
Many national facilities use Centos Linux and are unable or unwilling to make available the most recent stable version of <em>Emacs</em>.
I have found that the available versions are several years out of date and hence difficult to install all of my favorite packages.
It may take an hour of removing calls to install incompatible packages before you can establish a subset of your favorite packages.
Hence, it is good to have the ability to tolerate getting by for short periods of time with older versions of <em>Emacs</em> without all of your favorite packages.
</p>

<p>
Modes in <em>Emacs</em> are states of <em>Emacs</em> where a subset of commands are available.
Modes avoid clashes due to functions of the same name in different packages.
Modes are analogous to scopes in other text editors.
For example, there is a python-mode for working with Python script files.
</p>

<p>
Some modes are multilingual and have many additional non-editing functions.
One such mode is org-mode.
Org-mode was originally designed to be a planning and outlining tool.
Its ease of use for outlining lead to its extension to support literate programming in numerous programming languages. 
</p>

<p>
<em>Org-mode</em> has over 100 add-on packages.
For example, the org-ref package greatly eases the retrieval of pdfs, and the creation of BibTex entries.
The latter is done in an automated fashion by using the metadata in the pdf files.
</p>

<p>
The <em>.emacs.d</em> configuration directory is a hidden directory in the home directory.
It is the home of the installed packages and plugins.
</p>

<p>
The main configuration file resides is this directory and is named <em>init.el</em>.
The <em>init.el</em> file is the analog of Vim's <em>.vimrc</em> configuration file.
The commands in the <em>init.el</em> file are written in elisp.
Fortunately, <em>elisp</em> is easy to edit and augment without knowing much about how to program with <em>elisp</em>.
</p>

<p>
I store my configuration in a <em>config.el</em> file (see file listing at top of page).
</p>

<h3>Installing <em>Emacs</em></h3>
<p>
Like <em>Vim</em>, <em>Emacs</em> can be installed as a stand-a-lone application or via a software management tool and a software repository.
</p>
<p>
<em>Emacs</em> can be downloaded <a href="https://www.gnu.org/software/emacs/">from</a> and insalled with a platform specific installer.
Older versions of <em>Emacs</em> comes pre-installed on Mac OS, but you may want a more recent version.
You should remove the old version with the commands <code>sudo rm /usr/bin/emacs</code> and <code>sudo rm -rf /usr/share/emacs</code>.
Binaries for the Mac are <a href="https://emacsformacosx.com">available</a>.
As of August 2020, version 27.1-1 is the stable release.
</p>
<p>
<em>Emacs</em> is also available for a number of software repositories.
It can be installed on the Mac with macports using the command <code>sudo port install emacs </code>.
A GUI based version can be installed with the command <code>sudo port install emacs-app </code>.
The Homebrew repository for Mac OS is popular.
The brew command is <code>brew update && brew install emacs --with-cocoa && brew linkapps emacs </code>.
</p>
<p>
You may want to create an alias for starting <em>Emacs</em>.
For the development version of the binary in installed in Applications with the above package installer for the make use, <code>e28='open /Applications/Emacs2805.app $1'</code>.
For the app version installed by macports, use <code>e27='open -a /Applications/MacPorts/Emacs.app/Contents/MacOS/Emacs'</code>.
For Homebrew installed <em>Emacs</em>, use <code> alias emacs='/usr/local/Cellar/emacs/24.x/Emacs.app/Contents/MacOS/Emacs -nw' </code>.
</p>

<h3>Installing and using snippets with <em>Emacs</em></h3>
<p>
The package yasnippets is one of the most popular packages in <em>Emacs</em> for managing snippets.
This package provides for the display of the snippets in a table that opens in a separate buffer. 
The user can navigate the table to find a relevant snippet.
This table is redundant with the ls snippet.
</p>
<p>
The <a href="https://www.emacswiki.org/emacs/Yasnippet"><em>yasnippets</em></a> package is used to manage snippets in <em>Emacs</em>.
This package is installed with a built-in package manager.
</p>
<p>
Each snippet is stored in a single file.
The files are stored in a subfolder in the hidden folder <code>~.eamcs.d/plugins/</code> in the home directory.
</p>

<h1>Variants of <em>Emacs</em> to consider</h1>
<p>
These variants might be suitable for the impatient who do not want to master <em>GNU Emacs</em>.
</p>

<h2>Spacemacs</h2>
<p>
<a href="https://www.spacemacs.org">Spacemacs</a> (see below) is built ontop of GNU <em>Emacs</em>.
It is designed to be easier to use than <em>Emacs</em>.
It can be operated with <em>Vim</em>, <em>Emacs</em>, or a hybrid of key bindings.
You enable layers to extend its functionality.
<em>Spacemacs</em> can be used to edit <em>Jupyter Notebooks</em> via the Ipython-layer.
</p>

<h2><em>SciMax</em></h2>
<p>
<a href="http://kitchingroup.cheme.cmu.edu/scimax"><em>SciMax</em></a> is being developed by the chemical engineer John Kitchin at Carnegie Mellon University.
This variant of <em>Emacs</em> is being optimized for supporting the preparation of scientific manuscripts.
Several YouTube videos of Professor Kitchin talking about <em>SciMax</em> are available.
<em>SciMax</em> is on the to-be-added later list because the documentation for this project is lagging so the user has to be more self-reliant.
Hence, our recommendation to become competent with <em>Gnu Emacs</em> before diving into <em>SciMax</em>.
</p>

<h2><em>Prelude Emacs</em></h2>

<p>
This variant of <em>Emacs</em> is <em>GNU Emacs</em> bundled with what the developer thinks are the essential, no-frills add-on packages.
It does not ship with evil-mode. 
This variant might be appropriate for beginners.
You need to install <em>Gnu Emacs</em> version 26.3 first.
The <a href="https://prelude.emacsredux.com/en/latest/modules/python/">documentation </a> is on-line.
The GitHub page is found <a href="https://github.com/bbatsov/prelude">here</a>.
</p>

<h2><em>Doom Emacs</em></h2>

<p>
<em>Doom Emacs</em> appears to provide more advanced features than <em>Prelude Emacs</em>.
It may be useful to study to learn about packages or features to add to a basic <em>Gnu Emacs</em> configuration.
I learned about moving much of the code in the init.el file to code blocks in a config.org file, which can be rendered on GitHub like a markdown file.
The code blocks can be flanked by explanatory text, hyperlinks, tables, and figures.
The is a great way to document and share one's <em>Emacs</em> configuration.
</p>

<p>
<em>Emacs</em> reads an abbreviated init.el file which calls the <em>config.org</em> file.
The code in the code blocks in the config.org file are copied to a <em>config.el</em> file and then executed. 
</p>

<p>
See the file section at the top for an example of a working <em>init.el</em> and <em>config.org</em> file.
</p>
<a href=#FASTLINKS2>Return to list of editors above.</a>


<a name="espresso"><h2> <em>Espresso (Mac only)</em> </h2></a>
<p>
<a href="https://espressoapp.com/"><em>Espresso</em></a> is a proprietary test editor for the Mac.
It is for those users who desire a text editor that has a GUI that looks like a native Mac app.
The snippets are stored in a single <em>.xml</em> file.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name='geany'></a><h2> <em>Geany (Universal)</em></h2>

<p>
<a href="https://www.geany.org"><em>Geany</em></a> is a free and lightweight GUI-based text editor that has been under development since 2005.
<em>Geany</em> is written in C and C++.
Binaries are available for Mac OS, Linux, and Windows.
The source code is also available.
<em>Geany</em> has a plugin manager, and it is very easy to configure.
</p>

<h3>Installing snippets in Geany</h3>

<p>
This <a href="https://www.geany.org/download/releases/">webpage</a> has links to installers for each operating system.
</p>

<p>
Make sure that the <b>pml</b> filetype is defined as <em>#~ PyMOL=*.pml;</em> by going to <b>Tools &rarr; Configuration files &rarr; filetype_extensions.conf</b>.
You will also need to add PyMOL as a group in that same file configuration.
It should read <b>#~ None=PyMOL</b> under Groups.
</p>

<p align="center"><img src="gifs/geanyFiletypeConfig1st.gif"></p>

<p>
In addition, you need to edit the keybinding preferences <em>Move cursor in snippet</em> for under <b>Edit &rarr; Preferences &rarr; Keybindings</b> because it is blank by default.
I used <b>Tab</b>.
Press <b>enter</b> and then click <em>Allow</em> when the override window pops up.
Without making this edit, you will not be able to advance to the second and later tab stops.
</p>
<p>
The snippets for all languages are stored in groups in single file <em>~/.config/geany/snippets.conf</em>.
The snippets in this file can be edited by using the pulldown <b>Tools &rarr; Configuration files &rarr; snippets.conf</b>, but it might be easier to use a text editor.
</p>
<p>
Copy and paste the contents of the above <b>geanypymolsnippets.conf</b> file into the <b>snippets.conf</b> file.  
Make sure that there is only one section with the heading <b>[pymol]</b>.
This heading is case sensitive.
</p>

<h3>Using the snippet library in Geany</h3>
<p>
The snippets are invoked with a tab trigger.
The snippets have tab stops, but there is no support for the mirroring of tab stops.
</p>

<p>
Here is an example of the <b>ao</b> snippet being used in <em>Geany</em>:
</p>

<p align="center"><img src="gifs/geanyAoSnip.gif"></p>

A to-be-developed pml language file is needed to enable syntax highlighting of <b>pml</b> files.

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name='gedi'><h2 name="gedit"> Gedit (Universal) </h2></a>

<p>
<a href="https://wiki.gnome.org/Apps/Gedit"><em>Gedit</em></a> is the GUI-based text editor developed by the GNU project.
It is commonly found on Linux operating systems.
It is often available on the computers running Linux OS at national labs.
<em>Gedit</em> provides an uncluttered GUI with access to the snippets through a pulldown menu or by a tab trigger.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>


<h3>Installing and using snippets with <em>Gedit</em></h3>
<p>
The pymol.lang file has to be installed first.
Download it from the geditpymolsnips folder.
</p>
<pre>
	<code>
		mkdir -p ~/.local/share/gtksourceview-3.0/language-specs
	</code>
</pre>

<p>
The snippets are stored in a file called <em>pymol.xml</em>.
This file is stored in the home directory in a hidden folder called <em>.config</em>.
The full path is <code>~.config/gedit/snippets/pymol.xml</code>.
You may need to create this directory.
</p>

<p>
To do this:
</p>


<pre>
<code> 
	mkdir ~/.config/gedit/snippets/pymol.xml
</code>
</pre>

<p>
Then to copy <em>pymol.xml</em> into that directory, do:
</p>

<pre>
<code>
	cp ~/pymolsnips/geditpymolsnips/pymol.xml ~/.config/gedit/snippets/.
</code>
</pre>
	
<a href=#FASTLINKS2>Return to list of editors above.</a>


<a name='kate'><h2> Kate/Kwriter (Universal)</h2></a>
	
<a href=>"https://kate-editor.org/get-it/"><em>kate</em></a> is a free but KDE-dependent text editor for Linux that is now available for Mac and Windows.

<h3>Installing and using snippets with <em>Kate</em></h3>

<p>
If installed with MacPorts, the app resides in <code>/Applications/macports/kate.app</code>.
The snippet plugin has to be turned on.
<em>Kate</em> requires MacOS 10.14 or newer.
</p>

<p>
The snippets are stored in a single <em>.xml</em> file.
The storage location for this file is not in the documentation.
Try the following on Mac (last one of the four) and Linux (first three). 
<b>Be sure to change the folder from <em>YOURUSERNAME</em> to your username.</b>:
</p>

<pre>
<code>
/Users/YOURUSERNAME/.local/share/ktexteditor_snippets/data/pml-kate.xml
/Users/YOURUSERNAME/.kde4/share/apps/ktexteditor_snippets/data/pml-kate.xm
/Users/YOURUSERNAME/.kde/share/apps/ktexteditor_snippets/data/pml-kate.xm
/Users/YOURUSERNAME/Library/Application\ Support/kate/MySnippets/
</code>
</pre>
	
	
<p>
May have to create the directory MySnippets on MacOS.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="komodo"><h2 > Komodo Edit (Universal) </h2></a>
<p>
<a href="https://www.activestate.com/products/komodo-edit/"><em>Komodo Edit</em></a> is a proprietary program with a community version that can be used for free.
The developers of <em>Komodo Edit</em> have merged with the developers of ActiveState Python.
The community version of <em>Komodo Edit</em> has enough features for the occasional writer of code for PyMOL.
One fun aspect of <em>Komodo Edit</em> is the ability to use themes to customize your workspace.
<em>Komodo Edit</em> was derived from the Mozilla Framework.
The add-on, or extension, capabilities are the same.
For <em>Komodo Edit</em> documentation, see <a href="http://docs.activestate.com/komodo/11/">here</a>.
</p>

<h3>Installing and using snippets with <em>Komodo Edit</em></h3>
<p>
The snippets are treated as abbreviations and are stored as separate files in the tool box.
<em>Komodo Edit</em> will read the  <em>.pml</em> extension.
</p>

<p>
Snippets are stored in a directory called <em>PML</em>, which has to be created.
For Mac users,
</p>

<pre>
<code>
	mkdir ~/Library/Application\ Support/KomodoEdit/11.1/tools/Abbreviations/PML
</code>
</pre>

<p>
Then insert the snippets files from <a href="https://github.com/marevebro/pymolsnips/tree/master/komodoeditpymolsnips">here</a> into this directory.
</p>

<pre>
<code>
	$ cp -a ~/pymolsnips/komodoeditpymolsnips ~/Library/Application\ Support/KomodoEdit/11.1/tools/Abbreviations/PML
</code>
</pre>

<p>
For Windows users, select where you want <em>Komodo Edit</em> to be installed.
I chose <code>C:\Users\MYUSERNAME\AppData\KomodoEdit</code>.
From there, you'll need to create the <em>PML</em> directory in <code>...\KomodoEdit\lib\support\samples\tools\Abbreviations</code>.
Then copy and paste the <em>komodeditpymolsnips</em> files into the newly created <em>PML</em> directory. 
</p>

<p>
Here, you can see snippet use in action:
</p>

<p align="center"><img src="gifs/KomodoEditsnipuse.gif"></p>

<p>
The snippet name works as a tab trigger.
Unfortunately, snippets must be executed by memory as <em>Komodo Edit</em> does not have a suggestion window for snippets.
The snippet file can also be selected through the GUI.
A pop-up menu displays the option to insert a snippet.
There are tab stops and highlighted default parameter values.
However, there is no mirroring of tab stops at this time.
</p>
<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="LightTable"><h2> Light Table (Universal) </h2></a>

<p>	
<a href="http://lighttable.com/"><em>Light Table</em></a> is a highly malleable text editor that is written in ClojureScript, a programming language for the web that is a dialect of Lisp and that has a powerful macro system.
Those users who have dabbled in <em>Emacs</em> will recognize the syntax of the settings in behavior files which remind me of <em>Emacs</em> init files.
<em>Light Table</em> is like a GUI version of <em>Emacs</em>.
It has a very attractive, uncluttered GUI.
</p>

<p>
<em>Light Table</em> supports previews that enable interactive editing of code to change the output.
For example, the code chunk for generating a plot is evaluated in-line and the plot is displayed in an adjacent window.
The plugin manager can be used to install plugins, update them, and gain access to the documentation for the plugin.
This documentation is on a GitHub website.
</p>


<h3>Installing and using snippets with Light Table</h3>

<p>
The installation of plugins in <em>Light Table</em> is very fast.
The number of available plugins is not as extensive as for <em>Atom, Sublime Text, TextMate,</em> or <em>Visual Studio Code</em>.
You have to be prepared to do a bit of customization to get some of the plugins to work correctly.
</p>

<p>
The snippets are managed through one master file per programming language.
This central file has the file extension <em>.edn</em>.
This file is used to set the scope of the snippets (the programming language) and call the snippets, especially ones with multiple lines and indenting, from individual files with the file extension <em>.snip</em>.
</p>

<p>
The snippets are stored in <code>~/.lighttable/User/snippets</code>.
You will have to create this series of nested directories.
Enter the following command in bash:
</p>

<pre>
    <code>
	mkdir ~/.lighttable/User/snippets</code>
	</code>
</pre>

<p>
Download the <a href="https://github.com/MooersLab/pymolsnips/tree/master/lighttablepymolsnips"> <em>lighttablepymolsnips</em></a> folder from above.
Move the contents of this folder <b>(not the folder)</b> to <code>~/.lighttable/User/snippets</code>.
To do this:
</p>

<pre>
	<code>
	cp -a ~/pymolsnips/lighttablepymolsnips ~/.lighttable/User/snippets/.
	</code>
</pre>

<p>
Then, in <em>Light Table</em>, select from the menu in the toolbar at the top <b>File>Settings>User behaviors</b>.
This will open a new window with the <b>users.behaviors</b> file ready to be edited.
Paste inside the square brackets the following and change the path to your home directory:
</p>

<p>
This will open a new window with the <b>users.behaviors</b> file ready to be edited.
Paste inside the square brackets the following and change the path to your Mac home directory <b>be sure to change YOURUSERNAME to your name</b>:
</p>

<pre>
<code>
;; absolute path to where you would like your snippet root directory to be <br>
[:snippets.loader :lt.plugins.snippets.loader/set-snippet-dir <br>
"/Users/YOURUSERNAME/.lighttable/User/snippets"]
</code>
</pre>

<h3>For Windows</h3>

<p>
Save this file.
</p>

<p>
Download the <a href="https://github.com/MooersLab/pymolsnips#cudatext"><em>lighttablepymolsnips</em></a> folder from above.
Move the contents of this folder (not the folder) to <code>~/.lighttable/User/snippets</code>.
</p>

<p>
Here is that in action!
</p>

<p align="center"><img src="gifs/LightTableUserBehavior.gif"></p>	


<p>
There is not a <em>pml</em> language available yet for <em>Light Table</em>.
Select the Python lexer. 
It will work well enough.
To invoke the Pytho lexer, use <em>.py</em> as your file extension while editing your script file in <em>Light Table</em>, and then save your script file with the <em>.pml</em> file extension.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="micro"><h2> Micro (Universal) </h2></a>

<p>
<a href="https://github.com/zyedidia/micro/releases/tag/v1.4.1"><em>Micro</em></a> is an advance on the terminal-based editor <em>nano</em>.
Nonetheless, <em>Micro</em> is easier to use than <em>Vim</em> or <em>Emacs</em>.
Like <em>nano</em>, it uses keybindings between keyboard shortcuts and commands to get work done.
The collection of keybinding is defined in a <em>json</em> file.
<a href="https://github.com/zyedidia/micro"><em>Micro</em></a> has binaries available for Mac, Windows, Linux, and three flavors of BSD.
</p>

<h3>Installing and using snippets with Micro</h3>

<p>
<em>Micro</em> has a number of plugins available including one for snippets.
You need to have <em>Node.js</em> installed.
</p>

<p>
On the Mac, it is best to use <em>Micro</em> with the free iTerm terminal emulator.
</p>

<p>
Enter <em>Micro</em> on the command line to start the editor.
<em>Control-E</em> opens the command line at the bottom of the terminal window.
A list of key bindings are found <a href="https://github.com/zyedidia/micro/blob/master/runtime/help/keybindings.md">here</a>.
</p>

<p>
Install the <em>snippets</em> plugin by entering <em>control-E</em> in micro to open a command panel, and then enter:
</p>

<pre>
<code>
plugin install snippets
</code>
</pre>

<p>
All of the snippets for one language are in a single file with the <em>.snippets</em> file extension.
In this case, the file is labeled <em>pml.snippets</em>.
</p>

<p>
The snippets are installed in the folder <code>.config/micro/plugins/snippets</code>.
In the <em>pymolsnips</em> directory, enter:
</p>

<pre>
<code>
cp pml.snippets ~/.config/micro/plugins/snippets/snippets/.
</code>
</pre>

<p>
Snippets for other program languages are found <a href="https://github.com/tommyshem/micro-snippets-plugin/tree/master/snippets">here</a>.
</p>

<ul>
<li><b>Alt-S</b> is mapped to the <b>snippet0insert</b> command which inserts the snippet specified by a tabtrigger name.</li>
<li><b>Alt-W</b> advances the cursor to the next tab stop.</li>
<li><b>Alt-A</b> ends the editing the snippet.</li>
<li><b>Alt-D</b> deletes the snippet that is being edited.</li>
</ul>

<p>
A lexer for syntax highlighting is in the file called <em>pml.yaml</em>.
This file can be downloaded from the <a href="https://github.com/MooersLab/pymolsnips/tree/master/micropymolsnips"><em>micropymolsnips</em></a> folder above.
This lexer is stored in the hidden folder <code>.config/micro/syntax</code>, which you will need to create.
</p>

<pre>
<code>
mkdir ~/.config/micro/syntax
</code>
</pre>


<p>
Then in the <em>micropymolsnips</em> directory:
</p>

<pre>
<code>
cp pml.yaml ~/.config/micro/syntax/.
</code>
</pre>

<p>
Enter <b>Control-Q</b> and then enter <code>help color</code> to learn about the available color schemes and to learn about syntax highlighting.
I entered <code>set colorscheme bubblegum</code> to apply a light color scheme to <em>Micro</em> as shown below.
A PyMOL script file has been opened and is color-highlighted.
You can make the tex larger for easier reading by entering <code>cmd-+</code> a number of times.
</p>

<p align="center"><img src="./images/microBubblrGumColorScheme.png" width="640" height="400" alt="colorscheme"/></p>

<a href=#FASTLINKS2>Return to list of editors above.</a>


<a name='Neovim'><h2>Neovim</h2></a>
This a recent branch of Vim.
Much of the  <a href=#Vim>Vim</a> section applies.



<a name="spacemacs"><h2>Spacemacs (Universal)</h2></a>

<p>
Spacemacs (pronounced <em>space macs</em>) is a <em>Vim</em> emulator that runs on top of <em>Emacs</em>.
Its command leader is the spacebar hence the word <em>space</em> in the name.
It is very different from running the <em>Emacs</em> in evil-mode (evil-mode enables the use of Vim keybindings in <em>Emacs</em>).
As a result, it should be treated as a separate editor from <em>Emacs</em> and <em>Vim</em>.
</p>

<p>
<em>Spacemacs</em> has its own set of mnemonic commands that you need to master.
In other words, you will use <em>Vim</em>, <em>Emacs</em>, and <em>Spacemacs</em> commands in one editor.
It might be less intimidating if you have mastered some of the basics of <em>Vim</em> and <em>Emacs</em>.
<em>Spacemacs</em> does not go as far as allowing you to use Vim plugins and configurations although you do have full access to the emacs plugins which are known as packages.
</p>

<h3>Installing and using snippets with Spacemacs</h3>

<p>
<em>Spacemacs</em> comes with the package yasnippets installed, which is the engine for managing snippets.
<em>Spacemacs</em> use a layer concept to build up a specific configuration.
The layers are managed by editing the <b>*.spacemacs</b> file.
A few suggested layers are listed under <em>dotspacemacs-configuration-layers</em>.
Most of them are initially commented out with two semicolons on the left.
</p>

<p>
Add <em>yasnippet-snippets</em> to one of the defualt layers that are listed but are commented out with two semicolons on the left.
Remove the two semicolons.
Take care to use spaces and not a tab to indent yasnippet.
</p>

<p>
The snippets are installed by the <em>yasnippets-snippets</em> package.
You can use the following command in <em>Spacemacs</em> to install this package.
</p>

<pre>
<code>
M-x package-install RETURM
yasnippet
</code>
</pre>

<p>
<em>Spacemacs</em> may be more attractive to <em>vi</em> users rather than <em>Vim</em> users who cannot part with their configuration in <em>Vim</em>.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="SublimeText3"><h2>Sublime Text 3 (Universal)</h2></a>

<p>	
<a href="https://www.sublimetext.com/3"><em>Sublime Text 3</em></a> is a GUI-based editor.
The free trial period is infinite.
<em>Sublime Text 3</em> starts up much faster than the other text editors for programmers.
<em>Sublime Text 3</em> has strong support for snippets and the autocompletion of tab triggers.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<h3>Installing and using snippets with Sublime Text 3</h3>

<p>
The snippets are stored in separate files with the file extension <em>.sublime-snippet</em>.
</p>

<p>
<em>Sublime Text 3</em> strongly supports writing in LaTeX.
It provides previews of figures and math equations from within a tex document before compiling it to pdf.
Documents can be compiled on the fly and the bug reporting is more helpful than in most platforms.
</p>

<p>
For both all users (Mac, Windows, Linux), go to <code>Preferences&rarr;Browse Packages...&rarr;User&rarr;snippets</code>.
You may need to create the subdirectory called <em>snippets</em>.
Copy and paste <em>st3pymolsnips</em> into the <em>snippets</em> folder.
</p>

<p>
For command line action on the Mac, create the <em>snippets</em> directory in the <em>Sublime Text3</em> folder: <code>mkdir ~/Library/Application\ Support/Sublime\ Text\ 3/Packages/User/snippets/</code>
</p>

<p>
Add the snippet files to the <em>Sublime Text 3</em> folder:
</p>

<pre>
<code>
cp -a ~/pymolsnips/st3pymolsnips ~/Library/Application\ Support/Sublime\ Text\ 3/Packages/User/snippets/
</code>
</pre>

<p>
You can get around making the snippets subdirectory and just move <em>pymolsnips</em> to <code>~/Library/Application\ Support/Sublime\ Text\ 3/Packages/User/</code>, and it should still work.
</p>

<p>
Restart <em>Sublime Text 3</em>, open a <em>PyMOL</em> script file, and then enter <b>ao</b> and hit tab.
Sixteen lines of code should appear.
</p>


<p align="center"><img src="gifs/SiblimeAoSnip.gif"></p>

<p>
<em>Sublime Text 3</em> also supports tab stops and mirrored tab stops!
Below is an example of mirrored tab stop editing with the <b>threeMaps</b> snippet.
Notice how when I type <em>sugar</em> all instances of glycan are replaced with sugar.
</p>

<p align="center"><img src="gifs/SublimeMirror.gif"></p>

<p>
You may need to install a snippet manager package and add an autocompletion package to be able to get autocompletion of the tab triggers and a description of the snippet.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="TextMate"><h2>TextMate (Mac only)</h2></a>


<p>	
<a href="https://macromates.com"><em>TextMate</em></a> is freely available but only for Mac OS.
It is a mature project with occasional updates.
There is a large library of plugins available.
I found it to be quite intuitive to use, fast to start-up, and very robust.
It is the editor that I use most frequently.
</p>

<p>	
The 2007 book <a href="https://www.amazon.com/Textmate-James-Edward-II-Gray/dp/097873923X"><em>TextMate Power Editing for the Mac</em></a> by Edward Gray and the 2012 book <a href="https://www.amazon.com/TextMate-How-Chris-Mears-ebook/dp/B009XB55PW/ref=sr_1_2?dchild=1&keywords=TextMate+How-To&qid=1606051948&s=books&sr=1-2"><em>TextMate How-To</em></a> by Chris Mears provide supplemental reading to the on-line documentation.
</p>

<p>	
<em>TextMate</em>'s approach to snippets has served as a role model for other text editors.
<em>TextMate</em> snippets can be ported to <em>Sublime Text 3</em> and <em>PyCharm</em>.
Chapter 6 in the book by Edward Gray provides a good overview of the features of snippets in <em>TextMate</em>.
</p>

<h3>Installing and using snippets with TextMate</h3>

<p>	
Each snippet is stored in a separate file with the file extension <em>.tmSnippet</em>.
Each snippet file has a unique uuid string.
The snippet files are stored in a subfolder called <em>Snippets</em>.
This folder is stored in a folder <em>PyMOL.tmbundle</em>.
This folder can be dragged and dropped into the user's <code>~/Library/Application\ Support/TextMate/Bundles</code> folder.
When installing an update of the library, it is best to delete the old <em>PyMOL.tmbundle</em> and then drop in the new <em>PyMOL.tmbundle</em>.
</p>

<p>
The <em>PyMOL.tmbundle</em> folder contains several files and folders in addition to the <em>Snippets</em> folder. 
Some of this other files include a lexer for <em>PyMOL</em>.
The one file is the <em>dependencies.json</em> file.
This file specifies the dependence on the pygments package.
</p>

<p>
The second file is the <em>info.plist</em> file.
This file has the developer's contact information, the name of the snippet collection for the <em>Bundles</em> pull-down menu in <em>TextMate</em>, and a unique uuid string.
</p>

<p>
The <em>PyMOL.tmbundle</em>  shows up in the Bundles pull-down as <em>PyMOL.pml</em>.
Select <em>PyMOL.pml</em> and then a new menu will appear with a list of the snippets.
The descriptions of the snippets are on the left, and the corresponding tab trigger is on the right.
The snippet can be added to the current <em>.pml</em> document by selecting the snippet with the GUI or by entering the tab-trigger name.
</p>


<p>
The bundle pull-down provides access to a bundle editor.
This editor can be used to view and change the code in the snippets.
</p>
<p align="center"><img src="https://media.giphy.com/media/dvO0a4s2sArYieymKr/giphy.gif" width="640" height="480" alt="ao tab trigger"/></p>

<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name='Vim'><h2>Vim (Universal)</h2></a>

<p>
<a href="https://vimawesome.com"><em>Vim</em></a> is the improved terminal-based text editor <em>vi</em>.
<em>Vi</em> is very ubquitious.
It is probably already installed on your computer.
<em>Vim</em> enables the addition of plugins to extend what the base <em>Vim</em> can do.
There are over 19,000 <a href="https://vimawesome.com/"><em>plugins available</em></a>.
<em>Vim</em> is free.
</p>

<p>
A newer branch of <em>Vim</em>called <em>neovim</em> is also available.
<em>Vim</em> and <em>neovim</em> are available from software managers like MacPorts, fink, Homebrew, Anaconda, or it can be downloaded as a <a href="https://www.vim.org/download.php"> stand-alone application</a>.
The standard versions of both run in the terminal.
MacVim provides a GIU version of <em>Vim</em>.
This version of <em>Vim</em> is integrated with Mac OS.
Unless you a developer of GUI applications for the Mac, you probably want to stick to the terminal version
</p>

<p>
It is useful to be conversant with several of the above package managers because one of them is bound to have a working version of <em>Vim</em>.
<em>Vim</em>  version 8.1.0 was not working on Mac OS X Catalina due to a bug in <em>Xcode</em>.
That bug has been fixed in version 8.1.2152.
I found that the MacPorts Vim was behind this version.
I tried to compile <em>Vim</em>  version 8.2.12; I gave up due to inference with the conda in the <em>Julia</em> installed by <em>Homebrew</em>.
So I gave up my approach using macports, and I installed Vim version 8.2.0 via <em>Homebrew</em>.
This version worked fine.
It has been automatically updated to 8.2.1950 when running <em>brew update</em>.
</p>

<p>
If you have <a href="https://anaconda.org/"><em>Anaconda</em></a> or <a href="https://docs.conda.io/en/latest/miniconda.html"><em>Miniconda</em></a> installed, you can install <em>Vim</em> with conda as follows: <code>conda install -c conda-forge vim</code>.
</p>

<p>
<em>Vim</em> focuses on the editing process.
This process is where you spend most of your time thinking about what you have written.
Most of the time spent <em>writing</em> is usually spent on <em>rewriting</em>.
</p>

<p>
The <em>Vim</em> keybindings for text editing are so useful that have been ported to many <a href="https://www.reddit.com/r/vim/comments/3tluqr/my_list_of_applications_with_vi_keybindings/"><em>other applications</em></a> where the text is edited.
For example, <em>Vim</em> key bindings are available for the rival text editors <em>Emacs</em> as well as many other text editors including <em>VSC</em> and even in the on-line LaTeX editor <a href="https://www.overleaf"> <em>Overleaf</em></a>.
There is a <em>Vim</em> keybinding extension for <em>MS Word</em> on Windows but not on Mac OS.
This is one of many reasons why I do most of my writing in <em>Overleaf</em>.
</p>

<p>
The secret to keeping your sanity while using <em>Vim</em> is to remember to escape from the insert mode to the normal mode as soon as you finish entering a chunk of text because the normal mode is where you issue editing commands.
Numerous blog posts have lists of ways to think about and use <em>Vim</em> to improve your productivity.
The consensus is to develop muscle memory of the commands that are useful to you by practicing their use at every opportunity.
Another tip is to focus on mastering the basic commands before getting mired in tweaking your <em>.vimrc</em> configuration file and playing with plugins; that is, focus on mastering the <em>vi</em> part of <em>Vim</em>.
</p>

<p>
A common trap for new users is to overload their <em>.vimrc</em> configuration file with shiny new features and plugins that they find on GitHub from posts of sample <em>.vimrc</em> files.
Some of these files are several thousand lines long and really only usable by their developers.
New users often accidentally add one line that breaks <em>Vim</em>, get frustrated, and give up on <em>Vim</em>
A better practice is to slowly and irregularly build-up the <em>.vimrc</em> file as your needs evolve.
You can comment out features that you use infrequently and uncomment them as the need arises.
Obviosly, your <em>.vimrc</em> file should be kept under version control so that you can easily recover the last working version.
</p>

<p>
<em>Vim</em> lacks native support for snippets.
We have to extend <em>Vim</em> with a plugin known as a snippet manager.
There are several plugins for managing snippets.
<em>Neosnippets</em>, <em>SnipMate</em>, and <em>UltiSnips</em> will be discussed here.
</p>

<p>
Versions of Vim 8.0 and greater can use minpac to manage plugins on top of the native support for plugins.
Versions of <em>Vim</em> 7.X and earlier require the use of one of the plugin managers like Vundle to install plugins.
Most plugins have GitHub websites of their own.
These sites rarely mention minpac, but Vundle is always mentioned.
</p>

<p>
<em>Vundle</em>, which stands for <em>Vim bundle</em>, is a plugin manager that configures your plugins directly in the <em>.vimrc</em> file.
It is in this file that you install and update plugins, as well as comment out unused plugins.
</p>


<a href=#FASTLINKS2>Return to list of editors above.</a>



<b>Installing Vundle</b>

<p>
To start adding a plugin to <em>Vim</em> you will need to open and edit the <em>.vimrc</em> file.
To download Vundle enter the following into the bash command line:
</p>

<pre>
<code>
git clone https://github.com/VundleVim/Vundle.vim.git ~/.vim/bundle/Vundle.vim
</code>
</pre>

<p>
Paste the following minimal code at the top of your <em>.vimrc</em> file to allow the use of Vundle:
</p>



<pre>
	<code>
if has('python3')
endif

vundle
set nocompatible filetype off
set rtp+=~/.vim/bundle/Vundle.vim

" Add only comment lines and Plugin commands between
" call vundle#begin() and call vundle#end()

call vundle#begin()
plugin 'VundleVim/Vundle.vim'

" Ultisnips requires Vim installed with python. This is just the engine.
Plugin 'SirVer/ultisnips'
call vundle#end()

" Turn on filetype detection for plugins
filetype plugin on

" set path to Python3 interpreter that you used when installing Vim
" let g:python3_host_prog = '/opt/local/bin/python3.7'
" This was the python used to install Vim with homebrew.
let g:python3_host_prog = '/usr/local/Cellar/python/3.7.5/bin/python3'

" UltiSnips related commands
let g:UltiSnipsExpandTrigger = ''
let g:UltiSnipsJumpForwardTrigger = ''
let g:UltiSnipsJumpBackwardTrigger = ''
let g:UltiSnipsSnippetDirectories=[$HOME.'.vim/my-snippets/Ultisnips'
	</code>
</pre>

<p>
Now open <em>Vim</em> and run the following command:  <code>:PluginInstall</code>
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>


<h3>Minpac</h3>

<p>
<em>Minpac</em> is anther a plugin manager for installing and managing plugins in <em>Vim</em>.
It requires <em>Vim</em> 8 or newer.
Unfortunately, minpac does not update itself.
Fortunately, there is a package you can add that will allow minpac to auto-update.
This is discussed below in the minpac installation section.
</p>

<h3>Installing minpac</h3>

<p>	
To install minpac, you need to create a directory called <em>minpac</em> and install it as an optional plugin:
</p>

<pre>
<code>
mkdir -p ~/.vim/pack/minpac/opt
cd ~/.vim/pack/minpac/opt
git clone https://github.com/k-takata/minpac.git
</code>
</pre>

<p>
Then, add the following to your <em>.vimrc</em> file and write the file <b>:w</b>:
</p>

<pre>
<code>
packadd minpac
call minpac#init()
</code>
</pre>

<p>
To add a plugin using minpac you will need to add this command <code>call minpac#add('author/reponame')</code> to your <em>.vimrc</em> file.
For example, if you wanted to install snippets for <em>UltiSnips</em> you would add (this will not work at the moment) to the <em>.vimrc</em> file:
</p>

<pre>
<code>
call minpac#add('MooersLab/pml_bhmm.snippets')
</code>
</pre>

<p>
To allow minipac to update itself enter:
</p>

<pre>
<code>
call minpac#add('k-takata/minpac',{'type':'opt'})
</code>
</pre>

<p>
Then refresh <em>Vim</em> giving the <code>:source %</code> command and update minpac plugins with <code>:call minpac#update()</code>.
You will not be able to use these plugins until you exit and restart <em>Vim</em> by entering <code>:qa!</code> and then opening another <em>Vim</em> session.
</p>


<a name="neosnippets"><h3>Neosnippets(for <em>Vim</em>, universal) </h3></a>

<p>
<a href="https://github.com/Shougo/neosnippet.vim"><em>Neosnippets</em></a> is a snippet management plugin for the text editors <em>Vim</em> and <em>neovim</em>.
<em>Neosnippets</em> is similar to <em>SnipMate</em> except with <em>Neosnippets</em>, there is the option to utilize <a href="https://github.com/Shougo/deoplete.nvim"><em>deoplete</em></a> interface to suggest in-line alternate snippets.
The use of <em>Neosnippets</em> does not require <em>deoplete</em>.
Experienced users may find <em>deoplete</em> pop-up menus to be distracting.
<em>Deoplete</em> may be useful as a training tool that can be dispensable when it is no longer needed.
<em>Deoplete</em> requires <em>neovim</em> or <em>Vim</em> versions greater than 8.0, and Python3.
</p>

<a href=#FASTLINKS2>Return to list of editors above.</a>


<h4>Installing and using snippets with Neosnippets</h4>

<p>
You have to enable the Python3 interface with pynvim which you can install with this command:
</p>

<pre>
<code>
pip3 install --user pynvim
</code>
</pre>

<p>
To be more precise, you can provide the full path to the Python interpreter to be used by <em>neovim</em> and import <em>pip</em> as a module before installing pynvim.
</p>


<pre>
<code>
sudo -H /opt/local/bin/python3.7 -m pip install pynvim
</code>
</pre>

<p>
Check in <em>Vim</em> or <em>neovim</em> if python3 is enabled by entering:
</p>

<pre>
<code>
:echo has("python3")
</code>
</pre>

<p>
If 1 is returned, you have python3 installed.
If 0 is returned, you do not.
</p>

<p>
To install the plugins with Vundle, add the following to the Vundle plugin section of the <em>.vimrc</em> file.
</p>

<pre>
<code>
Plugin 'Shougo/deoplete.nvim'
Plugin 'roxma/nvim-yarp'
</code>
</pre>

<p>
To use <em>deoplete</em>, add the following setting to your <em>.vimrc</em> file.
</p>

<pre>
<code>
" Use deoplete.
let g:deoplete#enable_at_startup = 1
</code>
</pre>

<p>
To install <em>Neosnippets</em>, add the following to the Vundle plugin section of the  <em> .vimrc</em> file.
</p>

<pre>
<code>
Plugin 'Shougo/neosnippet.vim'
Plugin 'Shougo/neosnippet-snippets'
</code>
</pre>

The last command is optional because installing default snippets is optional.
If you choose not to install them, you must deactivate them with the following command added to the <em> .vimrc</em> file.

<code>g:neosnippet#disable_runtime_snippets</code>

<p>
A minimal <em> .vimrc</em> file for installing and using <em>Neosnippets</em> is listed below. 
Make sure to change YOURUSERNAME to your name:
</p>


<pre>
<code>
if has('python3')
endif

"vundle
set nocompatible
filetype off
set rtp+=~/.vim/bundle/Vundle.vim

" Add only comment lines and Plugin commands between
" call vundle#begin() and call vundle#end()
call vundle#begin()

Plugin 'VundleVim/Vundle.vim'

" neosnippets uses deoplete to suggest snippets
" This might be superior to ultisnips.
Plugin 'roxma/nvim-yarp'
Plugin 'roxma/vim-hug-neovim-rpc'
Plugin 'Shougo/neosnippet.vim'
Plugin 'Shougo/neosnippet-snippets'

call vundle#end()

" Turn on filetype detection for plugins
filetype plugin on

let g:python3_host_prog = '/opt/local/bin/python3.7'
let g:pydiction_location = '/Users/YOURUSERNAME/.vim/bundle/pydiction/complete-dict'

" Use deoplete.
let g:deoplete#enable_at_startup = 1

" maximum width of deoplete window
call deoplete#custom#source('_', 'max_menu_width', 80)


" i for swichting to the insert mode from normal mode;
" ii for escape from insert mode to normal mode
:inoremap ii  <Esc>

" another means of escape to normal mode: Control-Carriage Return
:inoremap <C-CR>  <Esc>
</code>
</pre>


<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="snipmate"><h3> SnipMate (for <em>Vim</em>, universal) </h3></a>

<p>
<a href="https://github.com/garbas/vim-snipmate"><em>SnipMate</em></a> is an early, but stable plugin manager that provides plugin support for snippets that are available in <em>Vim</em>. 
It does not depend on Python.
</p>

<h4>Installing and using snippets with SnipMate</h4>

<p>
To install SnipMate, add the following to your `.vimrc` file:
</p>

<pre>
<code>
Plugin 'MarcWeber/vim-addon-mw-utils'
Plugin 'tomtom/tlib_vim'
Plugin 'garbas/vim-snipmate'

" Optional:
Plugin 'honza/vim-snippets'
</code>
</pre>

<p>
Then run the following command in <em>Vim</em>:  <code>:PluginInstall</code>.
</p>


<a href=#FASTLINKS2>Return to list of editors above.</a>



<a name="UltiSnips"><h3> UltiSnips (for <em>Vim</em>, universal) </h3></a>

<p>
<a href="https://github.com/SirVer/ultisnips"><em>UltiSnips</em></a> is a more recent plugin manager that depends on Python3.
<em>Vim</em> has to be installed with the option of using Python3 enabled to be able to use <em>UltiSnips</em>.
The <em>UltiSnips</em> plugin is the snippet handling engine.
It provides a large number of advanced snippet features.
However, it does not come with libraries of snippets.
The snippets must be installed separately.
The <em>honza/vim-snippets</em> plugin contains snippets for many languages.
The <em>pml.snippets</em> file can be stored in a <em>mysnippets</em> subfolder of the hidden <em>.vim</em> directory as described below.
</p>

<em>UltinSnips</em> can read <em>SnipMate</em> snippets.



<h4>Installing and using snippets with UltiSnips</h4>
<p>
In the following setup, we use the Vundle plugin manager.

The <em>PyMOL.snippets</em> could be stored inside the <em>UltiSnips</em> directory, but they would be lost if you delete the <em>UltiSnips</em> directory.
To help make sure your snippets are not deleted if the <em>UltiSnips</em> directory ever gets deleted you will want to create a subdirectory called <em>myultisnips</em> at the top level of the <em> .vim</em> directory.
Move the <em>PyMOL.snippets</em> file to this subdirectory.
We will inform <em>Vim</em> of this location when customizing the <em> .vimrc</em> file for <em>UltiSnips</em> a little further down.
</p>


<pre>
<code>
$ cd .vim
$ mkdir myultisnips
</pre>
</code>

In the <em>myultisnips</em> directory, you can add your own snippets and add the <em>UltiSnips</em> snippets provided above in the <em>ultisnippymolsnips</em> folder.

<pre>
<code>
$ cp ~/pymolsnips/ultisnippymolsnips/* .
</pre>
</code>

<p>
Because <em>UltiSnips</em> depends on Python3, you may have to install Python3 to be able to use it.
You can point <em>UltiSnips</em> to a particular Python3 interpreter with the following command in your <em>.vimrc</em> file.
At the top of the file, add:
</p>

<pre>
<code>
if has('python3')
endif
</pre>
</code>

<p>
In the Vundle plugin section of the <em>.vimrc</em> file, add the following lines:
</p>

<pre>
<code>
" Ultisnips requires <em>Vim</em> installed with python.
Plugin 'SirVer/ultisnips'
</pre>
</code>

<p>
Below the plugin section, add the following:
</p>

<pre>
<code>
" Turn on filetype detection for plugins
filetype plugin on

let g:python3_host_prog = expand('/opt/local/bin/python3.7')

" UltiSnips related commands
let g:UltiSnipsExpandTrigger = '<tab>'
let g:UltiSnipsJumpForwardTrigger = '<tab>'
let g:UltiSnipsJumpBackwardTrigger = '<s-tab>'
let g:UltiSnipsSnippetDirectories=['/Users/blaine/.vim/my-snippets/Ultisnips','UltiSnips']
</pre>
</code>

<p>
The following step is to define the <em>*.pml</em> file type inside a <em>pml.vim</em> file.
The <em>pml.vim</em> file is stored inside of a subfolder that contains the file type definition files.
This subfolder is called <em>ftdetect</em>, which is stored inside of the <em>~/.vim/bundle/ultisnips</em> folder.
This step is not necessary if you install the <em>bioSyntax-vim</em> plugin because it contains a <em>pml.vim</em> in its ftdetect subfolder.
The <em>bioSyntax-vim</em> also provides color syntax highlighting for <em>.pml</em> files.
</p>

<p>
Make a `ftdetect` directory at the top level of the `.vim` directory.
</p>

<pre>
<code>
$ mkdir -p ~/.vim/ftdetect/
$ cd ~/.vim/ftdetect/
</pre>
</code>

<p>
Some <em>Vim</em> plugins require that you add a flag to your <em>.vimrc</em> file to turn off file type detection.
You can set the file type for a <em>.pml</em> file with the command <em>:set filetype=pml</em>.
To enable autodetection of the file type in this situation, we need to make a <em>pml.vim</em> file in the <em>ftdetect</em> subdirectory of the <em>ultisnips</em> directory.
</p>


<pre>
<code>
$ cd ~/.vim/bundle/ultisnips/ftdetect/
$ touch PyMOL.vim
$*Vim*PyMOL.vim
</pre>
</code>

<p>
Insert the single-line code fragment below into the file <em>pml.vim</em> and save it.
Press <em>i</em> to get into insert mode.
Then <em>escape</em> to return to normal mode.
To save and quit type <em>:wq</em> on the command line of normal mode.
</p>


<pre>
<code>
au BufNewFile,BufRead *.pml set filetype=pml
</pre>
</code>

<p>
Now, make a soft link to the <em>ftdetect</em> directory inside the <em>ultisnips</em> subdirectory to this new directory.
</p>

<pre>
<code>
ln -s ~/.vim/bundle/ultisnips/ftdetect/* ~/.vim/ftdetect/
</pre>
</code>

<p>
To test your setup, open a PyMOL script file.
Enter on the command line in <em>Vim</em> <em>:set filetype?</em>.
You should bet back <em>filetype=pml</em>.
</p>

<p>
Here is a minimal, no-frills <em>vimrc</em> file that uses Vundle and works with the <em>pml.snippets</em> file.
You may have to edit the path to the python3 binary.
</p>


<pre>
<code>
if has('python3')
endif

"vundle
set nocompatible
filetype off
set rtp+=~/.vim/bundle/Vundle.vim

" Add only comment lines and Plugin commands between
" call vundle#begin() and call vundle#end()
call vundle#begin()

Plugin 'VundleVim/Vundle.vim'

" Ultisnips requires <em>Vim</em> installed with python. This is just the engine.
Plugin 'SirVer/ultisnips'

call vundle#end()

" Turn on filetype detection for plugins
filetype plugin on

let g:python3_host_prog = expand('/opt/local/bin/python3.7')

" UltiSnips related commands
let g:UltiSnipsExpandTrigger = '<tab>'
let g:UltiSnipsJumpForwardTrigger = '<tab>'
let g:UltiSnipsJumpBackwardTrigger = '<s-tab>'
let g:UltiSnipsSnippetDirectories=['/Users/blaine/.vim/my-snippets/Ultisnips','UltiSnips']

" i for swichting to the insert mode from normal mode;
" ii for escape from insert mode to normal mode
:inoremap ii  <Esc>

" another means of escape to normal mode: Control-Carriage Return
:inoremap <C-CR>  <Esc>
</code>
</pre>

<h4> More information about UtliSnips </h4>

<p>
There are a number of videos on YouTube about using UltiSnips.
The developer of UltiSnips, SirVer (aka Holger Rapp) has a series of four videos that you are highly recommended.
The links to these videos are found in the README.md file of the <a href="https://github.com/sirver/UltiSnips"><em>UltiSnips GitHub Repository</em></a>.
However, for a faster introduction to UltiSnips, I recommend watching the video <a href="https://www.youtube.com/watch?v=F9Wss9GkF9Q&feature=emb_logo"><em>How I Use the UltiSnips VIM Plugin To Make Better Snippets</em></a> by Bryan Jenks.
</p>

<p>
If you are an experienced snippet user and want to add intelliSense-like autocompletion of the snippet names in <em>Vim</em>, take a look at <a href="https://github.com/neoclide/coc.nvim"><em>Conquerer of Completion</em></a> by Qiming Zhao.
They claim that their package will bring a <em>VSC</em>-like experience to <em>Vim</em>.
</p>
	
<a href=#FASTLINKS2>Return to list of editors above.</a>


<a name="VisualStudioCode"><h3> Visual Studio Code (Universal)</h3></a>

<p>
<a href="https://code.visualstudio.com"><em>Visual Studio Code</em></a> (VSC) is a free text editor that was recently bought by Microsoft.
Its start-up speed is between that of <em>Sublime Text 3</em> and <em>Atom</em>.
It is still under exponentially rapid development.
It is easily enhanced with extensions.
There is even a Grammarly extesnion!
<em>VSC</em> has strong support for snippets and for autocompletion with intelliSense.
It is fair to say that <em>VSC</em> is one of the most advanced text editors.
</p>

<p>
The installation of <em>VSC</em> extensions is very painless and very fast.
There is an extension marketplace that is accessible from the left column within the <em>VSC</em> GUI.
You can search the marketplace for extensions.
Each extension has its own HTML  file with information about the extension.
This HTML file has <b>install</b> and <b>uninstall</b> buttons.
The installations are lightning-fast.
</p>

<p>
The <em>bioSyntax</em> plugin is free and includes a lexer for PyMOL and lexers for many other file types used in bioinformatics.
It is a good idea to install this plugin.
The PyMOL lexer will be activated when a <em>pml</em> file is opened.
</p>

<p>
<em>VSC</em> provides for autocompletion of the snippet tab triggers and previews of the snippet's code.
The latter feature is very helpful for avoiding the insertion of the wrong snippet.
</p>


<b>Installing and using snippets with Visual Studio Code</b>

<p>
Windows users will need to install the 'Remote - WSL' extension for the windows/Linux system.
A prompt should pop-up during your first launch of <em>VSC</em>. Install this.
</p>

<p>
From the above folder <em>vscpymolsnips</em>, download and move the file <em>pml.json</em> to <code>~/Library/Application\ Support/Code/User/snippets</code> on the Mac; on windows,<code>$HOME/APPDATA\Roaming\Code\User\snippets\pml.json</code>; on Linux, <code>$HOME/.config/Code/User/snippets/pml.json</code>.
</p>

<p>
For Mac users in the <em>vscpymolsnips</em> folder:
</p>


<pre>
<code>
mv pml.json ~/Library/Application\ Support/Code/User/snippets
</code>
</pre>

<p>
For Windows users, select the settings button (gear icon) in the bottom left corner. 
Go to <em>Settings &rarr; User Snippets &rarr; pml</em>.
If you already have a <em>pml</em> file for snippets, simply append the <em>pml.json</em> file to the existing file (by copying and pasting).
</p>

<p>
Next, install <em>bioSyntax</em> via <em>View &rarr; Extensions</em>.
Search for <em>bioSyntax</em> and hit the <b>Install</b> button.
You will be asked to select a theme, just select bioSyntax.
</p>


<p align="center"><img src="gifs/VSCbioSyntax.gif"></p>

<p>
Open a file with a <em>.pml</em> file extension.
You should see <em>pml</em> in the lower right on the bottom panel.
Type the tab trigger <b>ao</b> to try inserting the snippet for the ambient occlusion effect.
You should see syntax highlighting.
</p>

<p align="center"><img src="gifs/VSCaoSnip2.gif"></p>

<p>
<em>VSC</em> also supports the use of mirrored tab stops.
In the example shown below, I am using the snippet <em>threeMaps</em>.
</p>

<p align="center"><img src="gifs/VSCMirror.gif"></p>

<a href=#FASTLINKS2>Return to list of editors above.</a>

<hr>

<a name="Snippet Lists"><h2>List of the snippets by their tab trigger name and description</h2></a>

<a name='Alternate locators'> <h2>Alternate locators:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                  </th></tr>

</thead>
<tbody>
<tr><td>hidealtloc   </td><td>Hide the b alternate locator.</td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Change orientation'> <h2>Change orientation:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                        </th></tr>
</thead>
<tbody>
<tr><td>rotate       </td><td>Rotate a selection about and axis by a given angle.</td></tr>
<tr><td>rv           </td><td>Return settings in rounded format.                 </td></tr>
<tr><td>turnAboutAxis</td><td>Turn about axis.                                   </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Color scheme'> <h2>Color scheme:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger          </th><th WIDTH="200">Description                                                                                                                                                                                                                                                                                                                                                                       </th></tr>
</thead>
<tbody>
<tr><td>cblind               </td><td>Apply color blind friendly to ribbon diagrams.                                                                                                                                                                                                                                                                                                                                    </td></tr>
<tr><td>cribbon              </td><td>Color ribbon H red, strand yellow, loop green.                                                                                                                                                                                                                                                                                                                                    </td></tr>
<tr><td>grayscale            </td><td>Apply grayscale coloring using a grayscale version of the PyMOL colors for the elements. This is a Python function. It is invoked in a script file via grayscale(). There is a corresponding gscale shortcut in pymolshortcuts.py that is invoked in a pml script by entering gsale if the functions in pymolshortcuts.py have been loaded with the run pymolshortcuts.py command.</td></tr>
<tr><td>printColorByAtomCodes</td><td>Print the codes for color-by-atom (util.cba*) alternates.                                                                                                                                                                                                                                                                                                                         </td></tr>
<tr><td>volumeRamp           </td><td>Volume ramp.                                                                                                                                                                                                                                                                                                                                                                      </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Electron density'> <h2>Electron density:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger         </th><th WIDTH="200">Description                                          </th></tr>
</thead>
<tbody>
<tr><td>carvedIsomesh       </td><td>Carved isomesh representation of electron density.   </td></tr>
<tr><td>carvedIsosurface    </td><td>Carved isosurface representation of electron density.</td></tr>
<tr><td>carvedVolume        </td><td>Carved volume representation of electron density.    </td></tr>
<tr><td>fetch2FoFcIsomesh   </td><td>Fetch 2FoFc map as an isomesh.                       </td></tr>
<tr><td>fetch2FoFcIsosurface</td><td>Fetch 2FoFc map as an isosurface.                    </td></tr>
<tr><td>fetch2FoFcVolume    </td><td>Fetch 2FoFc map as a volume.                         </td></tr>
<tr><td>fetchThreeMaps      </td><td>Display three electron density maos as isomesh.      </td></tr>
<tr><td>loadThreeMaps       </td><td>Three electron density as Isomesh.                   </td></tr>
<tr><td>threeMapsIsosurface </td><td>Three electron density as isosurface.                </td></tr>
<tr><td>threeMapsVolume     </td><td>Three electron density as volumes.                   </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='FileIO'> <h2>FileIO:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                               </th></tr>
</thead>
<tbody>
<tr><td>fetchCIF     </td><td>Fetch cif file.                                           </td></tr>
<tr><td>fetchFoFc    </td><td>Fetch fofc map.                                           </td></tr>
<tr><td>loadPDBfile  </td><td>Load a pdb file in the current directory.                 </td></tr>
<tr><td>printPath    </td><td>Print the path to the currently used PyMOL binary.        </td></tr>
<tr><td>saln         </td><td>Save an aln flle with a timestamp.                        </td></tr>
<tr><td>salnpy       </td><td>Save an aln flle with a timestamp. Python version.        </td></tr>
<tr><td>savePNG      </td><td>Save a png file of current scene to the current directory.</td></tr>
<tr><td>sccp4        </td><td>Save electron density map flle with timestamp             </td></tr>
<tr><td>sccp4py      </td><td>Save electron density map flle with timestamp             </td></tr>
<tr><td>sdae         </td><td>Save dae flle with timestamp                              </td></tr>
<tr><td>sdaepy       </td><td>Save dae flle with timestamp                              </td></tr>
<tr><td>spng         </td><td>Save png flle with timestamp                              </td></tr>
<tr><td>spngpy       </td><td>Save png flle with timestamp                              </td></tr>
<tr><td>spse         </td><td>Save pse flle with timestamp                              </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='H-bonds'> <h2>H-bonds:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                         </th></tr>
</thead>
<tbody>
<tr><td>distance     </td><td>H-bond distance between a H-bond donor and acceptor.</td></tr>
<tr><td>drawHbonds   </td><td>Display H-bonds as dashes colored black.            </td></tr>
<tr><td>hbonddash    </td><td>Set up H-bond dashes.                               </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Help'> <h2>Help:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger               </th><th WIDTH="200">Description                                                                                                                                                          </th></tr>
</thead>
<tbody>
<tr><td>cmddocs                   </td><td>Print in the command history window the docstrings of all of the functions in the cmd module.                                                                        </td></tr>
<tr><td>github                    </td><td>Print url of README.md file of the pymolsnips repository.                                                                                                            </td></tr>
<tr><td>helpDocs                  </td><td>Return the docustring for the help submodule. This command is more concise:  help help.                                                                              </td></tr>
<tr><td>presetDocs                </td><td>Return the docustring for the preset submodule. The command help preset fails to return anything. The command help(pymol.preset) has the same effect as help(preset).</td></tr>
<tr><td>printDoc                  </td><td>Print document string of a function.                                                                                                                                 </td></tr>
<tr><td>printDocpy                </td><td>Print document string of a function.                                                                                                                                 </td></tr>
<tr><td>pymoldocs                 </td><td>Return to the command history window the docstrings for all of the functions in a module.                                                                            </td></tr>
<tr><td>pymoldocspy               </td><td>Return to the command history window the docstrings for all of the functions in a module.                                                                            </td></tr>
<tr><td>writeCommandReference2HTML</td><td>Write the command reference to html file in the present working directory.                                                                                           </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Jupyter'> <h2>Jupyter:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger         </th><th WIDTH="200">Description                                                                                                                                                             </th></tr>
</thead>
<tbody>
<tr><td>AO                  </td><td>Run the AO function from the pymolshortcuts.py file to generate the photorealistic effect.                                                                              </td></tr>
<tr><td>AOBW                </td><td>Run the AOBW function from the pymolshortcuts.py file to generate photorealistic effect in grayscale.                                                                   </td></tr>
<tr><td>AOD                 </td><td>Run the AOD function from the pymolshortcuts.py file to generate photorealistic effect with carbons colored black.                                                      </td></tr>
<tr><td>AODBW               </td><td>Run the AODBW function from the pymolshortcuts.py file to generate photorealistic effect with carbons colored black and all other atoms colored in grayscale.           </td></tr>
<tr><td>PE125               </td><td>Run the PE125 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.125 of the van der Waals surface.                     </td></tr>
<tr><td>PE25                </td><td>Run the PE25 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.25 of the van der Waals surface.                       </td></tr>
<tr><td>PE33                </td><td>Run the PE33 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.33 of the van der Waals surface.                       </td></tr>
<tr><td>PE50                </td><td>Run the PE50 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.50 of the van der Waals surface.                       </td></tr>
<tr><td>PE66                </td><td>Run the PE66 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.66 of the van der Waals surface.                       </td></tr>
<tr><td>PE75                </td><td>Run the PE75 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.75 of the van der Waals surface.                       </td></tr>
<tr><td>PE85                </td><td>Run the PE85 function from the pymolshortcuts.py file to show the pearl effect with the inner sphere scaled by 0.85 of the van der Waals surface.                       </td></tr>
<tr><td>cav                 </td><td>Run the cav function from the pymolshortcuts.py file to show buried cavities and pockets as molecular surfaces.                                                         </td></tr>
<tr><td>cntccp4emaps        </td><td>Count number of *.ccp4 (electron density map) files in current directory.                                                                                               </td></tr>
<tr><td>cntfiles            </td><td>Count number of files in current directory.                                                                                                                             </td></tr>
<tr><td>cntlogs             </td><td>Count number of *.log files in current directory.                                                                                                                       </td></tr>
<tr><td>cntmtzs             </td><td>Count number of *.mtz files in current directory.                                                                                                                       </td></tr>
<tr><td>cntpdbs             </td><td>Count number of *.pdb files in current directory.                                                                                                                       </td></tr>
<tr><td>cntpmls             </td><td>Count number of *.pml files in current directory.                                                                                                                       </td></tr>
<tr><td>cntpngs             </td><td>Count number of *.png files in current directory.                                                                                                                       </td></tr>
<tr><td>cntpses             </td><td>Count number of *.pse files in current directory.                                                                                                                       </td></tr>
<tr><td>colorh1             </td><td>Run the colorh1 function from the pymolshortcuts.py file to color protein molecules according to the Eisenberg hydrophobicity scale, scheme 1.                          </td></tr>
<tr><td>colorh2             </td><td>Run the colorh2 function from the pymolshortcuts.py file to color protein molecules according to the Eisenberg hydrophobicity scale, scheme 2.                          </td></tr>
<tr><td>hb                  </td><td>Creates an object of all H-bonds found by PyMOL.                                                                                                                        </td></tr>
<tr><td>importIPythonDisplay</td><td>Imports for using IPython to display images loaded from disk in notebook cells.                                                                                         </td></tr>
<tr><td>importPyMOLcmd      </td><td>Import the cmd class from the pymol api.                                                                                                                                </td></tr>
<tr><td>importShortcuts     </td><td>Import for loading the functions in the pymolshortcuts.py file. These functions can be run inside cmd.do() without the trailing (). For example, cmd.do('rv')..         </td></tr>
<tr><td>imports4PyMOLjupyter</td><td>Imports needed for most uses of pymol in Jupyter. Combination of importPyMOL and importPythonDisplay.                                                                   </td></tr>
<tr><td>ipymolProtein       </td><td>Demo of the use of the RPC server with a protein via ipymol.                                                                                                            </td></tr>
<tr><td>ipymolStart         </td><td>Code to start the RPC server with ipymol.                                                                                                                               </td></tr>
<tr><td>kernel              </td><td>A kernel.json file for runnig PyMOL python interpreter in the Jupyter notebook. This code should reside in a folder named pymol.python in the ~/Library/Jupyter/kernels.</td></tr>
<tr><td>loadImage           </td><td>Load image.                                                                                                                                                             </td></tr>
<tr><td>nmr                 </td><td>Show all models in a nmr structure.                                                                                                                                     </td></tr>
<tr><td>nmroff              </td><td>Hide all but first model in a nmr structure.                                                                                                                            </td></tr>
<tr><td>nmroffpy            </td><td>Hide all but the first model in a nmr structure.                                                                                                                        </td></tr>
<tr><td>nmrpy               </td><td>Show all models in a nmr structure.                                                                                                                                     </td></tr>
<tr><td>rdkrpcChem          </td><td>Demo of the use of the RPC server with a drug compound via the rdkit python module.                                                                                     </td></tr>
<tr><td>rdkrpcProtein       </td><td>Demo of the use of the RPC server with a protein via rdkit.                                                                                                             </td></tr>
<tr><td>rmd                 </td><td>Remove all measurement objects in the interal GUI.                                                                                                                      </td></tr>
<tr><td>rmhb                </td><td>Delete all H-bonds in the selection, which is all by default.                                                                                                           </td></tr>
<tr><td>rv                  </td><td>Run the rv function from the pymolshortcuts.py file.                                                                                                                    </td></tr>
<tr><td>rvi                 </td><td>Return settings in rounded format while running PyMOL via the RCP server ipymol in a jupyter notebook. This is a modified version of the roundview.py script.           </td></tr>
<tr><td>rvr                 </td><td>Return settings in rounded format while running PyMOL via the RCP server rdkit in a jupyter notebook. This is a modified version of the roundview.py script.            </td></tr>
<tr><td>timcolor            </td><td>Run the timcolor function from the pymolshortcuts.py file to color atoms accordings to Tim Mather's biophysical coloring scheme for proteins..                          </td></tr>
<tr><td>yrb                 </td><td>Run the yrb function from the pymolshortcuts.py file.                                                                                                                   </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Labels'> <h2>Labels:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger   </th><th WIDTH="200">Description                                                                                                                                                                                          </th></tr>
</thead>
<tbody>
<tr><td>centerpi      </td><td>Center pi.                                                                                                                                                                                           </td></tr>
<tr><td>displayFonts  </td><td>Print to the screen as labels the 21 font ids in their corresponding fonts in a grid. Each label is an object and appears in the internal gui. You can turn on and off the display of specific fonts.</td></tr>
<tr><td>displayFontspy</td><td>Print to the screen as labels the 21 font ids in their corresponding fonts in a grid. Each label is an object and appears in the internal gui. You can turn on and off the display of specific fonts.</td></tr>
<tr><td>labelCAs      </td><td>Label the CA atoms with the Ala333 style format.                                                                                                                                                     </td></tr>
<tr><td>labelMainChain</td><td>Label the main chain atoms with the following: resn,resi,atom name.                                                                                                                                  </td></tr>
<tr><td>labelResnResi </td><td>Label CA atom with single-letter residue name and residue number.                                                                                                                                    </td></tr>
<tr><td>labelSS       </td><td>Label SS.                                                                                                                                                                                            </td></tr>
<tr><td>labelWatersHOH</td><td>Label waters with HOH and their residue number.                                                                                                                                                      </td></tr>
<tr><td>labelWatersW  </td><td>Label waters with W and their reisude number.                                                                                                                                                        </td></tr>
<tr><td>oneLetter     </td><td>Switch from three letter code to one-letter code for amino acids.                                                                                                                                    </td></tr>
<tr><td>pseudolabel   </td><td>Position label with pseudoatom.                                                                                                                                                                      </td></tr>
<tr><td>sigang        </td><td>Set angle labels to display 2 decimals places                                                                                                                                                        </td></tr>
<tr><td>sigdihedral   </td><td>Set dihedral labels to display 2 decimals places to the right of the decimal point.                                                                                                                  </td></tr>
<tr><td>sigdist       </td><td>set distance labels to display 2 decimals                                                                                                                                                            </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='List Snippets'> <h2>List Snippets:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                  </th></tr>
</thead>
<tbody>
<tr><td>lsSnips      </td><td>List all snips by tab trigger and description</td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Measurements'> <h2>Measurements:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger               </th><th WIDTH="200">Description                                                                      </th></tr>
</thead>
<tbody>
<tr><td>listLigandProteinDistances</td><td>Print a list of protein--ligand distances. Code by Dan Kulp. Updated for Python3.</td></tr>
<tr><td>ms                        </td><td>Measure surface area of the selection with the msms_pymol.py script.             </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Molecular representation'> <h2>Molecular representation:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger                   </th><th WIDTH="200">Description                                                                                                                                                                                                                                                                                                                                                                    </th></tr>
</thead>
<tbody>
<tr><td>ao                            </td><td>Apply the ambient occlussion effect to get the photorealistic effect.                                                                                                                                                                                                                                                                                                          </td></tr>
<tr><td>aobw                          </td><td>Ambient occlussion in grayscale.                                                                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>aod                           </td><td>Ambient occlussion with carbon atoms colored black.                                                                                                                                                                                                                                                                                                                            </td></tr>
<tr><td>aodbw                         </td><td>Ambient occlussion in grayscale with carbon atoms colored black.                                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>bs                            </td><td>Ball and stick representation.                                                                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>bu                            </td><td>Generate the biological unit using the quat.py script.                                                                                                                                                                                                                                                                                                                         </td></tr>
<tr><td>coordinate                    </td><td>Coordinate covalent bonds to metals and H-bonds from RNA.                                                                                                                                                                                                                                                                                                                      </td></tr>
<tr><td>cspheres                      </td><td>Colored spheres.                                                                                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>discreteCartoonColoring       </td><td>Turn on discrete colors between secondary structure elements.                                                                                                                                                                                                                                                                                                                  </td></tr>
<tr><td>doubleBond                    </td><td>Valence bond.                                                                                                                                                                                                                                                                                                                                                                  </td></tr>
<tr><td>drawLinks                     </td><td>Connect the alpha carbons of residue 1 with 10, 6 with 16, 7  with 17 and 8 with 18. Note that this example requires the draw_links.py [http://pldserver1.biochem.queensu.ca/~rlc/work/pymol/draw_links.py] by Robert Campbell.                                                                                                                                                </td></tr>
<tr><td>ellipcol                      </td><td>Set ellipsoid color.                                                                                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>filledRing                    </td><td>Filled rings in nucleic acids.                                                                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>fog                           </td><td>Blur the background atoms.                                                                                                                                                                                                                                                                                                                                                     </td></tr>
<tr><td>grayscalepy                   </td><td>Apply grayscale coloring using a grayscale version of the PyMOL colors for the elements. This is a Python function. It is invoked in a script file via gscale(). There is a corresponding gscale shortcut in pymolshortcuts.py that is invoked in a pml script by entering gsale if the functions in pymolshortcuts.py have been loaded with the run pymolshortcuts.py command.</td></tr>
<tr><td>his31asp70                    </td><td>Display the famous Asp70-His31 salt-bridge from T4 lysozyme that contributes3-5 kcal/mol to protein stability.                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>loadPDBbs                     </td><td>Load PDB ball-and-stick.                                                                                                                                                                                                                                                                                                                                                       </td></tr>
<tr><td>loadPDBnb                     </td><td>Load PDB nb spheres.                                                                                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>molscriptRibbon               </td><td>Show cartoon in the style of Molscript ribbons.                                                                                                                                                                                                                                                                                                                                </td></tr>
<tr><td>oneBondThicknessColor         </td><td>To change stick color and radius for the bond between atom 2 and 3, use the set_bond command.                                                                                                                                                                                                                                                                                  </td></tr>
<tr><td>pearl                         </td><td>The pearl effect is made with two spheres with the outer sphere being transparent.                                                                                                                                                                                                                                                                                             </td></tr>
<tr><td>puttyCartoon                  </td><td>Create a putty cartoon.                                                                                                                                                                                                                                                                                                                                                        </td></tr>
<tr><td>ringMode                      </td><td>Set the ring mode to a value between 0  and 6 in cartoons of nucleic acids.                                                                                                                                                                                                                                                                                                    </td></tr>
<tr><td>rmwater                       </td><td>Remove waters from molecular object.                                                                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>sas                           </td><td>Show the solvent excluded surface.                                                                                                                                                                                                                                                                                                                                             </td></tr>
<tr><td>saxsEnvelope                  </td><td>Display SAXS envelope                                                                                                                                                                                                                                                                                                                                                          </td></tr>
<tr><td>scaleRadiusColor              </td><td>Scale the radius and color of atoms as spheres by property in the B-value column.                                                                                                                                                                                                                                                                                              </td></tr>
<tr><td>scaleRadiusColorPythonInsertpy</td><td>Python block insert for scaleRadiusColorpy.                                                                                                                                                                                                                                                                                                                                    </td></tr>
<tr><td>scaleRadiusColorpy            </td><td>Scale the radius and color of atoms as spheres by property in the B-value column.                                                                                                                                                                                                                                                                                              </td></tr>
<tr><td>setLigandValenceOn            </td><td>Turn on the ligand valence.                                                                                                                                                                                                                                                                                                                                                    </td></tr>
<tr><td>setcolor                      </td><td>Set color name to a RGB code.                                                                                                                                                                                                                                                                                                                                                  </td></tr>
<tr><td>sidehChainHelper              </td><td>In cartoons, hide the backbone atoms of selected residues when showing then as sticks.                                                                                                                                                                                                                                                                                         </td></tr>
<tr><td>solventRadius                 </td><td>Set radius of ball used to make solvent accessible surface.                                                                                                                                                                                                                                                                                                                    </td></tr>
<tr><td>stack                         </td><td>Base-stacking figure.                                                                                                                                                                                                                                                                                                                                                          </td></tr>
<tr><td>stateOne                      </td><td>Select state 1 from a model with multiple states.                                                                                                                                                                                                                                                                                                                              </td></tr>
<tr><td>waterTriple                   </td><td>triple water pentagon.                                                                                                                                                                                                                                                                                                                                                         </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Nucleic Acids'> <h2>Nucleic Acids:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger                             </th><th WIDTH="200">Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  </th></tr>
</thead>
<tbody>
<tr><td>basePairStacking                        </td><td>This code make as standard base stacking diagram with ball and stick representation.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         </td></tr>
<tr><td>brokenNucleicBackbone                   </td><td>Create bonds between phosphorous and O3* atoms in a low-resolution DNA structure that is 80 base pairs long.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>dssrBlock1                              </td><td>Combining DSSR block representation with regular PyMOL cartoons after loading the dssr_block.py script by Thomas Holder.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     </td></tr>
<tr><td>dssrBlock2                              </td><td>DSSR block representation with fused blocks after loading the dssr_block.py script by Thomas Holder. The x3dna-dssr executable needs to be in the PATH.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      </td></tr>
<tr><td>dssrBlock3                              </td><td>DSSR block representation for a multi-state example after loading the dssr_block.py script by Thomas Holder. The x3dna-dssr executable needs to be in the PATH.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              </td></tr>
<tr><td>dssrBlock4                              </td><td>DSSR block representation with custom coloring after loading the dssr_block.py script by Thomas Holder. The x3dna-dssr executable needs to be in the PATH.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   </td></tr>
<tr><td>nucleicAcidBackboneTubesSticks          </td><td>This code shows the cartoon backbone tube as 65% transparent. It hides the rungs of the cartoon. It shows all of the non-H atoms are sticks. The motivation is to have the cartoon highlight the backbone without dominanting the scene.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     </td></tr>
<tr><td>nucleicAcidCartoon                      </td><td>Settings for nucliec acid cartoon.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>nucleicAcidCartoon2Strands              </td><td>Coloring two strand differently of a double helix makes it easier to for the viewer to distinguish the two strands. The set command has the syntax of setting_name [, setting_value [, selection [,state ]]] . In this case, the selection has to be global, object, object-state, or per-atom settings. It cannot be a named selection. This is a weak spot in PyMOL. Coloring two strand differently of a double helix makes it easier to for the viewer to distinguish the two strands. Many double-stranded helices have one strand in the asymmetric unit. The second strand is in the biological unit. The coordinates for the second strand are in the pdb1 file type at the PDB. The second strand is in the second state, which is equivalent to the second model in the pdb file. The strands are labeled chain A and B (via the cartoon_nucliec_acid_color setting). The bases are colored differently too (via the cartoon_ladder_color setting).</td></tr>
<tr><td>nucleicAcidCartoonFilledRings           </td><td>The code provides a cartoon of the loaded nucleic acid that has the ladder rungs replaced by filled rings that are colored by atom type. The code can be applied to any nucleic acid. It is derived from the FR shortuct in pymolshortcuts.py.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>nucleicAcidColorbySequence              </td><td>This code colors the nucleotides by base seqence. It can be applied to any nucleic acid.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     </td></tr>
<tr><td>nucleicAcidDumbellCartoonColorbySequence</td><td>This code colors by the nucleotides by base seqence. The backcone is shown as a flatten ribbon with rolled edges that give the dumbell effect. The code can be applied to any nucleic acid. The code is dervied from the CR and DU shortcuts.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                </td></tr>
<tr><td>nucleicAcidFlatRibbonColorbySequence    </td><td>This code colors the nucleotides by base seqence. It can be applied to any nucleic acid. It is dervied from the CR shortcut. The backcone is shown as a flatten ribbon.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Objects'> <h2>Objects:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger           </th><th WIDTH="200">Description                                                                                                                                                                                      </th></tr>
</thead>
<tbody>
<tr><td>allPairs              </td><td>This is a two-fold nested list comprehension for any all-parwise operation on the currently loaded objects. Replace optAlginRNA with any other function that operations on a pair of structrures.</td></tr>
<tr><td>listObjects           </td><td>Create a list of objects in the internal gui and print this list to the screen.                                                                                                                  </td></tr>
<tr><td>loadAndAlignManyFiles1</td><td>These are the instructions for loading and aligning multiple files. To save multiple models in a file to separate pdb files.                                                                     </td></tr>
<tr><td>loadAndAlignManyFiles2</td><td>To align all of the loaded RNA structures in all possible combinations by their C1' carbon atoms.                                                                                                </td></tr>
<tr><td>loadAndAlignManyFiles3</td><td>These are the instructions for loading and aligning multiple files.                                                                                                                              </td></tr>
<tr><td>loadManyFiles         </td><td>Load into PyMOL multiple files with a common file stem. The is a script by Robert Campbell that has been updated for Python3.                                                                    </td></tr>
<tr><td>optAlignRNA           </td><td>OptiAlign.py by Jason Vertree modified for aligning multiple RNA structures.                                                                                                                     </td></tr>
<tr><td>optAlignRNA           </td><td>OptiAlign.py by Jason Vertree modified for aligning multiple RNA structures.                                                                                                                     </td></tr>
<tr><td>saveSeppy             </td><td>Saves multiple objects into multiple files using an optional prefix name.                                                                                                                        </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Print'> <h2>Print:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger                </th><th WIDTH="200">Description                                                                                                                      </th></tr>
</thead>
<tbody>
<tr><td>aveB4resiX                 </td><td>AveBResiX, prints the residue number and the average bfactor.                                                                    </td></tr>
<tr><td>aveB4resiXpy               </td><td>AveBResiX, prints the residue number and the average bfactor.                                                                    </td></tr>
<tr><td>averageB                   </td><td>Average the B-factors by using a regular list as opposed to a stored list in PyMOL. Edit the selection as needed.                </td></tr>
<tr><td>fasta                      </td><td>Print Fasta from PDB file.                                                                                                       </td></tr>
<tr><td>fastapy                    </td><td>Python version of the command to print the sequence from a PDB file in the fasta format.                                         </td></tr>
<tr><td>findHbonds                 </td><td>Find H-bonds around a residue.                                                                                                   </td></tr>
<tr><td>getCoordinates             </td><td>Get coordinates.                                                                                                                 </td></tr>
<tr><td>getCoordinatespy           </td><td>Python version of getCoordinates snippets. Note that the python2 print statement stills works at the PyMOL prompt in pml scripts.</td></tr>
<tr><td>numResiNucleic             </td><td>Print the number of residues in a nulceic acid (all chains).                                                                     </td></tr>
<tr><td>numResiNucleicChainA       </td><td>Print the number of residues in a nulceic acid chain.                                                                            </td></tr>
<tr><td>numResiProtein             </td><td>Print the number of residues in a protein.                                                                                       </td></tr>
<tr><td>numResiProteinChainA       </td><td>Print the number of residues in chain A of a protein.                                                                            </td></tr>
<tr><td>printBs                    </td><td>Print the B-factors of a residue.                                                                                                </td></tr>
<tr><td>printBs2digits             </td><td>Print B--factors for a residue with the B-factors rounded off to two decimal places.                                             </td></tr>
<tr><td>printBspartB               </td><td>Print B factors of part B of a residue.                                                                                          </td></tr>
<tr><td>printNameB4ResiX           </td><td>Print name and b-factor for a residue.                                                                                           </td></tr>
<tr><td>printPathpy                </td><td>Print the path to the currently used PyMOL binary.                                                                               </td></tr>
<tr><td>printResiResnNameB4ResiX   </td><td>Print resn, resi, atom name, and b-factor.                                                                                       </td></tr>
<tr><td>printResiResnNameB4ResiXNoH</td><td>Print name and b-factor for a residue or residue range (e.g. 81:120). The noH variant.                                           </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Programming'> <h2>Programming:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger            </th><th WIDTH="200">Description                                                                                                </th></tr>
</thead>
<tbody>
<tr><td>emacsjupyterSourceBlock</td><td>Source block template in org-mode with emacs-jupyter package.                                              </td></tr>
<tr><td>obipythonSourceBlock   </td><td>Source block template in org-mode with the ob-ipython package.                                             </td></tr>
<tr><td>printAtomNames         </td><td>Print the atom names of a residue.                                                                         </td></tr>
<tr><td>printAtomNumbers       </td><td>Print the atom number2 of a residue.                                                                       </td></tr>
<tr><td>printBfactors          </td><td>Print the bfactors of a residue.                                                                           </td></tr>
<tr><td>printCoordinates       </td><td>Print the coordinates of the atoms in a residue.                                                           </td></tr>
<tr><td>printNamesCoordinates  </td><td>Print the atom names and coordinates of the atoms in a residue.                                            </td></tr>
<tr><td>printNamesCoordinates  </td><td>Print the atom names as tuples and coordinates of the atoms in a residue as a list.                        </td></tr>
<tr><td>printVDWradii          </td><td>Print the van der Waals radii of the atoms in of a residue.                                                </td></tr>
<tr><td>renameChain            </td><td>Rename a chain.                                                                                            </td></tr>
<tr><td>renumAtoms             </td><td>Add or substract a atom number offset.                                                                     </td></tr>
<tr><td>renumResi              </td><td>Add or substract a residue number offset.                                                                  </td></tr>
<tr><td>synch                  </td><td>Wait unitl all current commands have been executed. A timeout ensures that that command ecentually returns.</td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Pymolrc'> <h2>Pymolrc:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                              </th></tr>
</thead>
<tbody>
<tr><td>antialias    </td><td>Set antialias to on to get smoother edges.               </td></tr>
<tr><td>fetchPath    </td><td>Set path for location to save fetched pdb files.         </td></tr>
<tr><td>lspymolrc    </td><td>Print list of active pymolrc files.                      </td></tr>
<tr><td>lspymolrcpy  </td><td>Print list of active pymolrc files.                      </td></tr>
<tr><td>setpath      </td><td>Set additional path for PyMOL to search on startup       </td></tr>
<tr><td>sigDigits    </td><td>Set number of decimals places to show in distance labels.</td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Selection'> <h2>Selection:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger         </th><th WIDTH="200">Description                                                 </th></tr>
</thead>
<tbody>
<tr><td>duplicateObject     </td><td>Duplicate object.                                           </td></tr>
<tr><td>extractPartObj      </td><td>Create a new object from part of an existing object.        </td></tr>
<tr><td>hideSelection       </td><td>Turn off magenta squares on current selection.              </td></tr>
<tr><td>ligandSelect        </td><td>Make selection of ligand atoms.                             </td></tr>
<tr><td>selectAllBut        </td><td>Select all nitrogen atom in a selelction except from lysine.</td></tr>
<tr><td>selectAtomsAround   </td><td>Select atoms within a radius around a ligand.               </td></tr>
<tr><td>selectChain         </td><td>Select a chain.                                             </td></tr>
<tr><td>selectElement       </td><td>Select atoms by element.                                    </td></tr>
<tr><td>selectHelices       </td><td>Select atoms by alpha helices.                              </td></tr>
<tr><td>selectLoops         </td><td>Select atoms by beta loops.                                 </td></tr>
<tr><td>selectName          </td><td>Select atoms by name.                                       </td></tr>
<tr><td>selectResi          </td><td>Select residues by a range of numbers.                      </td></tr>
<tr><td>selectResidues      </td><td>Select residues by name.                                    </td></tr>
<tr><td>selectResiduesAround</td><td>Select residues within a radius around a ligand.            </td></tr>
<tr><td>selectStrands       </td><td>Select atoms by beta strands.                               </td></tr>
<tr><td>undoSelection       </td><td>Undo a selection.                                           </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Settings'> <h2>Settings:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                                                                           </th></tr>
</thead>
<tbody>
<tr><td>listSettings </td><td>Print to the screen the settings and their current parameter values. This is the more compact version.</td></tr>
<tr><td>listSettings2</td><td>Print to the screen the settings and their current parameter values.                                  </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Stereo'> <h2>Stereo:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger  </th><th WIDTH="200">Description                                   </th></tr>
</thead>
<tbody>
<tr><td>stereoDraw   </td><td>Stereo draw.                                  </td></tr>
<tr><td>stereoRay    </td><td>Stereo ray.                                   </td></tr>
<tr><td>stereokb     </td><td>Set keyboard shortcut by mapping F1 to stereo.</td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Trajectories'> <h2>Trajectories:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger   </th><th WIDTH="200">Description                                                   </th></tr>
</thead>
<tbody>
<tr><td>loadAmberTrajs</td><td>The amber trajectories have to be loaded into the same object.</td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Unit cell display'> <h2>Unit cell display:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger            </th><th WIDTH="200">Description                                                                                                                                                                                                                                                                                                                                                                                                    </th></tr>
</thead>
<tbody>
<tr><td>addAxis                </td><td>Adds the function draw_axis(). Used to a symmetry axis, a ncs axis, or scale bar to a scene.                                                                                                                                                                                                                                                                                                                   </td></tr>
<tr><td>addAxispy              </td><td>Adds the function draw_axis(). Could be useful for the adding a symmery axis, a ncs axis, or scale bar to a scene.                                                                                                                                                                                                                                                                                             </td></tr>
<tr><td>sc111                  </td><td>Display all symmetry mates in one unit cell. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>sc112                  </td><td>Display all symmetry mates in two unit cells along the c axis. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>sc113                  </td><td>Display all symmetry mates in three unit cels along c. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                       </td></tr>
<tr><td>sc114                  </td><td>Display all symmetry mates in four unit cells stacked long c-axis. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>sc121                  </td><td>Display all symmetry mates in two unit cells along the b axis. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>sc122                  </td><td>Display all symmetry mates in a 1 x 2 x 2 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>sc131                  </td><td>Display all symmetry mates in three unit cells along b. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                      </td></tr>
<tr><td>sc133                  </td><td>Display all symmetry mates in 1 x 3 x 3 array of unit cell. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                  </td></tr>
<tr><td>sc141                  </td><td>Display all symmetry mates in four unit cells stacked long b-axis. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>sc144                  </td><td>Display all symmetry mates in in a 1 x 4 x 4 array. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                          </td></tr>
<tr><td>sc211                  </td><td>Display all symmetry mates in two unit cell along a. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                         </td></tr>
<tr><td>sc212                  </td><td>Display all symmetry mates in a 2 x 1 x 2 arrays of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                              </td></tr>
<tr><td>sc221                  </td><td>Display all symmetry mates in 2 x 2 x 1 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>sc222                  </td><td>Run supercell script to generate three cells in all directions. This script was written by Thomas Holder.                                                                                                                                                                                                                                                                                                      </td></tr>
<tr><td>sc233                  </td><td>Display all symmetry mates in a 2 x 3 x 3 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>sc311                  </td><td>Display all symmetry mates three three unit cells along a. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                   </td></tr>
<tr><td>sc313                  </td><td>Display all symmetry mates in a 3 x 1 x 3 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>sc323                  </td><td>Display all symmetry mates in a 3 x 2 x 3 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                               </td></tr>
<tr><td>sc331                  </td><td>Display all symmetry mates in 3 x 3 x 1 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>sc332                  </td><td>Display all symmetry mates in 3 x 3 x 2 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>sc333                  </td><td>Display all symmetry mates in 3 x 3 x 3 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                 </td></tr>
<tr><td>sc411                  </td><td>Display all symmetry mates in four unit cells stacked long a-axis. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>sc414                  </td><td>Display all symmetry mates in a 4 x 1 x 4 array. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                             </td></tr>
<tr><td>sc441                  </td><td>Display all symmetry mates in four unit cells stacked long a-axis. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                           </td></tr>
<tr><td>sc444                  </td><td>Display all symmetry mates in a 4 x 4 x4 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                </td></tr>
<tr><td>sc444                  </td><td>Display all symmetry mates in a 4 x 4 x4 array of unit cells. Uses supercell.py in $HOME/Scripts/PyMOLscripts/.                                                                                                                                                                                                                                                                                                </td></tr>
<tr><td>symexp                 </td><td>The code expands the asymmetric unit. It like the generate symmetry mates command but it provides more control over the prefix name of the symmetry mates and the addition of unique segment identifiers for each symmetry mate. The usage: symexp prefix, object, (selection), cutoff, segidFlag. The cutoff is in Angstroms. The segidFlag set to 1 will add unique segids. For related functions, see SC***.</td></tr>
<tr><td>unitCellEdgesColorBlack</td><td>Color unit cell edges black. The settings for controlling the unit cell color are hard to find.                                                                                                                                                                                                                                                                                                                </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>


<a name='Workshop'> <h2>Workshop:</h2> </a>
<table class="table2">
<thead>
<tr><th WIDTH="80">Tab trigger     </th><th WIDTH="200">Description                                                            </th></tr>
</thead>
<tbody>
<tr><td>internalGUImode2</td><td>Make the background of the internal gui transparent to expand viewport.</td></tr>
<tr><td>internalGUIwidth</td><td>Set the width of the internal gui.                                     </td></tr>
</tbody>
</table>
<a href=#FASTLINKS3>Return to list of snippet categories near top.</a>
</HTML>
